// dc-delivery-sdk-js/build/main/lib/rendering/coordinators/RenderContentItem.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RenderContentItem = void 0;
var Url_1 = require("../../utils/Url");
var RenderedContentItem_1 = require("../model/RenderedContentItem");
var Edition_1 = require("../../content/model/Edition");
var ContentLifecycle_1 = require("../../content/model/ContentLifecycle");
/**
 * @hidden
 */
var RenderContentItem = /** @class */ (function () {
    function RenderContentItem(config, contentClient) {
        this.config = config;
        this.contentClient = contentClient;
    }
    RenderContentItem.prototype.renderContentItem = function (contentItemId, templateName, customParameters) {
        var _this = this;
        var queryParameters = this.getQueryParams(templateName, customParameters);
        var queryString = Url_1.encodeQueryString(queryParameters);
        var path = "/v1/content/" + encodeURIComponent(this.config.account) + "/content-item/" + encodeURIComponent(contentItemId);
        return this.contentClient.get(path + "?" + queryString).then(function (response) {
            return _this.parseResponse(response);
        });
    };
    RenderContentItem.prototype.getQueryParams = function (templateName, customParameters) {
        var queryParameters = [['template', templateName]];
        if (customParameters) {
            for (var _i = 0, _a = Object.keys(customParameters); _i < _a.length; _i++) {
                var key = _a[_i];
                var value = customParameters[key];
                queryParameters.push(["crparam." + key, value]);
            }
        }
        if (this.config.locale) {
            queryParameters.push(['locale', this.config.locale]);
        }
        return queryParameters;
    };
    RenderContentItem.prototype.parseResponse = function (response) {
        var headers = response.headers;
        var result = new RenderedContentItem_1.RenderedContentItem();
        result.body = response.data;
        if (headers) {
            var editionId = headers['X-Amp-Edition-ID'];
            var editionStart = headers['X-Amp-Edition-Start-Time'];
            var editionEnd = headers['X-Amp-Edition-End-Time'];
            var lifecycleExpiryTime = headers['X-Amp-Lifecycle-Expiry-Time'];
            if (editionId) {
                result.edition = new Edition_1.Edition({
                    id: editionId,
                    start: editionStart,
                    end: editionEnd,
                });
            }
            if (lifecycleExpiryTime) {
                result.lifecycle = new ContentLifecycle_1.ContentLifecycle({
                    expiryTime: lifecycleExpiryTime,
                });
            }
        }
        return result;
    };
    return RenderContentItem;
}());
exports.RenderContentItem = RenderContentItem;
