// dc-delivery-sdk-js/build/module/lib/content/mapper/ContentMapper.js
import { walkAndReplace } from '../../utils/JsonWalker';
import { FragmentMeta } from '../model/FragmentMeta';
import { ContentMeta } from '../model/ContentMeta';
import { Image } from '../../media/Image';
import { Video } from '../../media/Video';
import { ContentReference } from '../model/ContentReference';
/**
 * @hidden
 * ContentMapper provides functionality to convert simple JSON content into hydrated model classes
 * with helper functions. If a mapper is not defined for a schema it will remain as plain JSON.
 */
export class ContentMapper {
    constructor(config) {
        this.config = config;
        this.mappers = [];
        this.registerBuiltInMappers();
    }
    /**
     * Registers a custom model / fn to hydrate a specific schema
     * @param schema JSON schema ID or Regex that matches against the JSON schema ID
     * @param fn Model Class or Function that can convert the schema
     */
    addSchema(schema, fn) {
        if (typeof schema === 'string') {
            this.addCustomMapper((fragment) => {
                if (fragment._meta.schema === schema) {
                    return fn(fragment);
                }
            });
        }
        else {
            this.addCustomMapper((fragment) => {
                if (schema.test(fragment._meta.schema)) {
                    return fn(fragment);
                }
            });
        }
    }
    /**
     * Registers a custom function to hydrate content fragments
     * @param fn Mapper function
     */
    addCustomMapper(fn) {
        this.mappers.push(fn);
    }
    /**
     * Converts the provided content into hydrated model classes
     * @param content Content to convert
     */
    toMappedContent(content) {
        return walkAndReplace(content, {
            afterWalk: (node) => {
                if (FragmentMeta.isFragment(node)) {
                    return this.mapFragment(node);
                }
                return node;
            },
        });
    }
    /**
     * Converts a single fragment using the registered mappers
     * @param fragment Fragment to convert
     */
    mapFragment(fragment) {
        for (const mapper of this.mappers) {
            const result = mapper(fragment);
            if (result) {
                return result;
            }
        }
        return fragment;
    }
    /**
     * Registers built in mappers
     */
    registerBuiltInMappers() {
        this.addCustomMapper(this.convertContentReference.bind(this));
        this.addCustomMapper(this.convertContentMeta.bind(this));
        this.addCustomMapper(this.convertImage.bind(this));
        this.addCustomMapper(this.convertVideo.bind(this));
    }
    /**
     * Converts _meta inside Content Items into a ContentMeta instance
     * @param fragment
     */
    convertContentMeta(fragment) {
        if (ContentMeta.isContentBody(fragment)) {
            return {
                ...fragment,
                _meta: new ContentMeta(fragment._meta),
            };
        }
    }
    /**
     * Converts content-link into an Content reference class instance
     * @param fragment
     */
    convertContentReference(fragment) {
        if (ContentReference.isContentReference(fragment)) {
            const result = new ContentReference(fragment);
            return result;
        }
    }
    /**
     * Converts image-link into an Image class instance
     * @param fragment
     */
    convertImage(fragment) {
        if (Image.isImage(fragment)) {
            const result = new Image(fragment, this.config);
            return result;
        }
    }
    /**
     * Converts image-link into an Image class instance
     * @param fragment
     */
    convertVideo(fragment) {
        if (Video.isVideo(fragment)) {
            const result = new Video(fragment, this.config);
            return result;
        }
    }
}
