// dc-visualization-sdk/dist/lib/handler-container.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HandlerContainerFactory = exports.HandlerContainer = void 0;
var object_hash_1 = __importDefault(require("object-hash"));
var HandlerContainer = /** @class */ (function () {
    function HandlerContainer(event, connection) {
        this.event = event;
        this.connection = connection;
        this.changeHandlers = new Map();
    }
    HandlerContainer.prototype.add = function (cb, config) {
        var _this = this;
        var id = object_hash_1.default(config || {});
        if (this.changeHandlers.has(id)) {
            var existing = this.changeHandlers.get(id);
            var handlers = existing.add(cb);
            this.changeHandlers.set(id, handlers);
        }
        else {
            this.changeHandlers.set(id, new Set([cb]));
            if (this.event) {
                this.connection.emit(this.event, { id: id, config: config });
            }
        }
        return function () {
            var handlers = _this.changeHandlers.get(id);
            handlers === null || handlers === void 0 ? void 0 : handlers.delete(cb);
        };
    };
    HandlerContainer.prototype.run = function (model, handlerId) {
        var _a;
        if (handlerId === void 0) { handlerId = object_hash_1.default({}); }
        (_a = this.changeHandlers
            .get(handlerId)) === null || _a === void 0 ? void 0 : _a.forEach(function (handler) { return handler(model); });
    };
    return HandlerContainer;
}());
exports.HandlerContainer = HandlerContainer;
var HandlerContainerFactory = function (connection) { return function (event) {
    if (event === void 0) { event = null; }
    return new HandlerContainer(event, connection);
}; };
exports.HandlerContainerFactory = HandlerContainerFactory;
