// dc-visualization-sdk/dist/lib/settings.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Settings = exports.SETTINGS_EVENTS = exports.KEY = void 0;
var handler_container_1 = require("./handler-container");
exports.KEY = 'visualization-sdk:settings';
var SETTINGS_EVENTS;
(function (SETTINGS_EVENTS) {
    SETTINGS_EVENTS["GET"] = "visualization-sdk:settings:get";
    SETTINGS_EVENTS["CHANGE"] = "visualization-sdk:settings:change";
})(SETTINGS_EVENTS = exports.SETTINGS_EVENTS || (exports.SETTINGS_EVENTS = {}));
/**
 * Settings class allows you to get the currently selected visualization settings in the form such as Device selected, vse URL, contentId and snapshotId and watch for changes to asynchronously update your application
 */
var Settings = /** @class */ (function () {
    function Settings(connection) {
        var _this = this;
        this.connection = connection;
        var handlerContainer = handler_container_1.HandlerContainerFactory(connection);
        this.changeHandlerContainer = handlerContainer();
        this.connection.on(SETTINGS_EVENTS.CHANGE, function (settings) {
            _this.changeHandlerContainer.run(settings);
        });
    }
    /**
     * Get more context into the visualization, such as the vse and viewing device information
     *
     * ### Example
     *
     * ```typescript
     * const value = await visualization.settings.get();
     *
     * console.log(value)
     * ```
     */
    Settings.prototype.get = function () {
        return this.connection.request(SETTINGS_EVENTS.GET);
    };
    /**
     * Sets up a listener for when the visualization settings.
     *
     * @param cb - callback function to be called any settings change.
     *
     * @returns a dispose function which removes the listener
     *
     * ### Example
     *
     * ```typescript
     *  const dispose = visualization.settings.changed((model) => {
     *    // handle form change
     *  })
     * ```
     */
    Settings.prototype.changed = function (cb) {
        return this.changeHandlerContainer.add(cb);
    };
    return Settings;
}());
exports.Settings = Settings;
