// dc-delivery-sdk-js/build/module/lib/content/model/ContentMeta.js
import { Edition } from './Edition';
import { ContentLifecycle } from './ContentLifecycle';
import { FragmentMeta } from './FragmentMeta';
/**
 * Class providing meta data about an Content Reference resource.
 */
export class ContentReferenceMeta extends FragmentMeta {
    constructor(data) {
        super(data);
    }
}
/**
 * Class providing meta data about a content item with helper functions.
 */
export class ContentMeta extends FragmentMeta {
    /**
     * Creates a new ContentMeta instance.
     * @param data JSON representation of the ContentMeta model
     */
    constructor(data) {
        super(data);
        if (data) {
            if (data.edition) {
                this.edition = new Edition(data.edition);
            }
            if (data.lifecycle) {
                this.lifecycle = new ContentLifecycle(data.lifecycle);
            }
        }
    }
    /**
     * Export to JSON
     */
    toJSON() {
        const result = super.toJSON();
        if (this.deliveryId) {
            result.deliveryId = this.deliveryId;
        }
        if (this.deliveryKey) {
            result.deliveryKey = this.deliveryKey;
        }
        if (this.name) {
            result.name = this.name;
        }
        if (this.edition) {
            result.edition = this.edition.toJSON();
        }
        if (this.lifecycle) {
            result.lifecycle = this.lifecycle.toJSON();
        }
        if (this.hierarchy) {
            result.hierarchy = this.hierarchy;
        }
        return result;
    }
    /**
     * @hidden
     * Returns true if the provided node is a content item meta data object
     * @param node JSON node to test
     */
    static isContentMeta(node) {
        return (node != null &&
            node.schema != null &&
            (node.name != null || node.deliveryId != null));
    }
    /**
     * @hidden
     * Returns true if the provided node is a content body object
     * @param node JSON node to test
     */
    static isContentBody(node) {
        return node != null && this.isContentMeta(node._meta);
    }
}
