// dc-delivery-sdk-js/build/module/lib/rendering/coordinators/RenderContentItem.js
import { encodeQueryString } from '../../utils/Url';
import { RenderedContentItem } from '../model/RenderedContentItem';
import { Edition } from '../../content/model/Edition';
import { ContentLifecycle } from '../../content/model/ContentLifecycle';
/**
 * @hidden
 */
export class RenderContentItem {
    constructor(config, contentClient) {
        this.config = config;
        this.contentClient = contentClient;
    }
    renderContentItem(contentItemId, templateName, customParameters) {
        const queryParameters = this.getQueryParams(templateName, customParameters);
        const queryString = encodeQueryString(queryParameters);
        const path = `/v1/content/${encodeURIComponent(this.config.account)}/content-item/${encodeURIComponent(contentItemId)}`;
        return this.contentClient.get(`${path}?${queryString}`).then((response) => {
            return this.parseResponse(response);
        });
    }
    getQueryParams(templateName, customParameters) {
        const queryParameters = [['template', templateName]];
        if (customParameters) {
            for (const key of Object.keys(customParameters)) {
                const value = customParameters[key];
                queryParameters.push([`crparam.${key}`, value]);
            }
        }
        if (this.config.locale) {
            queryParameters.push(['locale', this.config.locale]);
        }
        return queryParameters;
    }
    parseResponse(response) {
        const headers = response.headers;
        const result = new RenderedContentItem();
        result.body = response.data;
        if (headers) {
            const editionId = headers['X-Amp-Edition-ID'];
            const editionStart = headers['X-Amp-Edition-Start-Time'];
            const editionEnd = headers['X-Amp-Edition-End-Time'];
            const lifecycleExpiryTime = headers['X-Amp-Lifecycle-Expiry-Time'];
            if (editionId) {
                result.edition = new Edition({
                    id: editionId,
                    start: editionStart,
                    end: editionEnd,
                });
            }
            if (lifecycleExpiryTime) {
                result.lifecycle = new ContentLifecycle({
                    expiryTime: lifecycleExpiryTime,
                });
            }
        }
        return result;
    }
}
