// dc-delivery-sdk-js/build/module/lib/media/Media.js
/**
 * @hidden
 * Base class for image or video resources.
 */
export class Media {
    constructor(data, config) {
        this.config = config;
        Object.assign(this, data);
    }
    /**
     * Returns the hostname that should be used to load this media resource.
     * The hostname will use the staging and media host overrides if specified
     * otherwise it will use the defaultHost provided by the delivery API.
     * @param secure
     */
    getHost(secure) {
        if (this.config.stagingEnvironment) {
            return this.config.stagingEnvironment;
        }
        if (secure) {
            return this.config.secureMediaHost || this.defaultHost;
        }
        else {
            return (this.config.mediaHost || this.config.secureMediaHost || this.defaultHost);
        }
    }
    /**
     * Export media object to JSON
     */
    toJSON() {
        return {
            defaultHost: this.defaultHost,
            endpoint: this.endpoint,
            name: this.name,
            id: this.id,
        };
    }
}
