// dc-management-sdk-js/build/main/lib/http/AxiosHttpClient.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AxiosHttpClient = void 0;
var axios_1 = __importDefault(require("axios"));
/**
 * @hidden
 */
var AxiosHttpClient = /** @class */ (function () {
    function AxiosHttpClient(config) {
        this.config = config;
        this.client = axios_1.default.create(config);
    }
    AxiosHttpClient.prototype.request = function (config) {
        return this.client
            .request({
            data: config.data,
            headers: config.headers,
            method: config.method,
            url: config.url,
        })
            .then(function (response) {
            return {
                data: response.data,
                status: response.status,
            };
        })
            .catch(function (error) {
            if (error && error.response) {
                return {
                    data: error.response.data,
                    status: error.response.status,
                };
            }
            return error;
        });
    };
    return AxiosHttpClient;
}());
exports.AxiosHttpClient = AxiosHttpClient;
