// dc-management-sdk-js/build/main/lib/hal/models/HalResource.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HalResource = void 0;
var HttpRequest_1 = require("../../http/HttpRequest");
/**
 * Base class for all Resources
 */
var HalResource = /** @class */ (function () {
    /**
     * Creates a new instance of the resource.
     * If data is provided it will be parsed as if it had
     * come from the remote api.
     * @param data
     */
    function HalResource(data) {
        if (data) {
            this.parse(data);
        }
    }
    /**
     * Parses the data returned by the API into the model class
     * @hidden
     */
    HalResource.prototype.parse = function (data) {
        Object.assign(this, data);
    };
    /**
     * Returns a copy of this resource's attributes excluding links and client references
     */
    HalResource.prototype.toJSON = function () {
        var result = Object.assign({}, this);
        delete result.client;
        delete result._links;
        delete result.related;
        return result;
    };
    /**
     * Returns a copy of this resource's attributes excluding links and client references
     * @deprecated
     */
    HalResource.prototype.toJson = function () {
        return this.toJSON();
    };
    /**
     * Set automatically by the HalClient when the resource is created.
     * If this is not set the resource will be unable to resolve related
     * resources and actions.
     *
     * @hidden
     */
    HalResource.prototype.setClient = function (client) {
        this.client = client;
    };
    /**
     * Converts an _embedded collection into an Array of
     * Resource instances.
     * @hidden
     */
    HalResource.prototype.parseEmbedded = function (name, resourceConstructor) {
        var _this = this;
        if (this._embedded[name]) {
            return this._embedded[name].map(function (x) {
                return _this.client.parse(x, resourceConstructor);
            });
        }
        else {
            return [];
        }
    };
    /**
     * GET a linked resource
     * @hidden
     */
    HalResource.prototype.fetchLinkedResource = function (name, params, resourceConstructor) {
        return this.withHalLink(name).then(function (_a) {
            var link = _a[0], client = _a[1];
            return client.fetchLinkedResource(link, params, resourceConstructor);
        });
    };
    /**
     * POST / CREATE a new resource
     * @hidden
     */
    HalResource.prototype.createLinkedResource = function (name, params, resource, resourceConstructor) {
        return this.withHalLink(name).then(function (_a) {
            var link = _a[0], client = _a[1];
            return client.createLinkedResource(link, params, resource, resourceConstructor);
        });
    };
    /**
     * POST to an action endpoint and get a resource response
     * @hidden
     */
    HalResource.prototype.performActionThatReturnsResource = function (name, params, data, resourceConstructor, method) {
        if (method === void 0) { method = HttpRequest_1.HttpMethod.POST; }
        return this.withHalLink(name).then(function (_a) {
            var link = _a[0], client = _a[1];
            return client.performActionThatReturnsResource(link, params, data, resourceConstructor, method);
        });
    };
    /**
     * POST to an action endpoint with no resource response returned.
     */
    HalResource.prototype.performActionWithoutResourceResponse = function (name, params, data, method) {
        if (method === void 0) { method = HttpRequest_1.HttpMethod.POST; }
        return this.withHalLink(name).then(function (_a) {
            var link = _a[0], client = _a[1];
            return client.performActionWithoutResourceResponse(link, params, data, method);
        });
    };
    /**
     * DELETE the current resource
     * @hidden
     */
    HalResource.prototype.deleteResource = function () {
        return this.deleteLinkedResource('delete', {});
    };
    /**
     * DELETE a linked resource
     * @hidden
     */
    HalResource.prototype.deleteLinkedResource = function (name, params) {
        return this.withHalLink(name).then(function (_a) {
            var link = _a[0], client = _a[1];
            return client.deleteLinkedResource(link, params);
        });
    };
    /**
     * PATCH the current resource
     * @hidden
     */
    HalResource.prototype.updateResource = function (resource, resourceConstructor) {
        return this.updateLinkedResource('update', {}, resource, resourceConstructor);
    };
    /**
     * PATCH a linked resource
     * @hidden
     */
    HalResource.prototype.updateLinkedResource = function (name, params, resource, resourceConstructor) {
        return this.withHalLink(name).then(function (_a) {
            var link = _a[0], client = _a[1];
            return client.updateLinkedResource(link, params, resource, resourceConstructor);
        });
    };
    HalResource.prototype.withHalLink = function (name) {
        if (!this.client) {
            return Promise.reject(new Error('HalResource has no client'));
        }
        var link = this._links[name];
        if (!link) {
            return Promise.reject("The " + name + " action is not available, ensure you have permission to perform this action.");
        }
        if (name === 'list-slots' && link.templated) {
            // Work around an API bug where pagination params are not included for list-slots.
            var toReplace = '{?includedSlots}';
            var toReplaceWith = '{?includedSlots,page,size,sort}';
            if (link.href.endsWith(toReplace)) {
                link.href =
                    link.href.substr(0, link.href.length - toReplace.length) +
                        toReplaceWith;
            }
        }
        return Promise.resolve([link, this.client]);
    };
    return HalResource;
}());
exports.HalResource = HalResource;
