// dc-management-sdk-js/build/main/lib/model/AssignedContentType.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.AssignedContentTypePage = exports.AssignedContentType = void 0;
var HalResource_1 = require("../hal/models/HalResource");
var Page_1 = require("./Page");
var Webhook_1 = require("./Webhook");
/**
 * Class representing an Assigned Content Type for an Algolia Search Index.
 */
var AssignedContentType = /** @class */ (function (_super) {
    __extends(AssignedContentType, _super);
    function AssignedContentType() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /**
         * Resources and actions related to an Assigned Content Type.
         */
        _this.related = {
            /**
             * Unassigns the content type from the search index.
             */
            unassign: function (id) {
                return _this.deleteLinkedResource('unassign', {
                    id: id,
                });
            },
            /**
             * Recreates the webhooks for this Assigned Content Type.
             */
            recreateWebhook: function () {
                return _this.performActionWithoutResourceResponse('recreate-webhook', {}, {});
            },
            /**
             * Gets the primary webhook for this Assigned Content Type.
             */
            webhook: function () {
                return _this.fetchLinkedResource('webhook', {}, Webhook_1.Webhook);
            },
            /**
             * Gets the active content webhook for this Assigned Content Type.
             */
            activeContentWebhook: function () {
                return _this.fetchLinkedResource('active-content-webhook', {}, Webhook_1.Webhook);
            },
            /**
             * Gets the archived content webhook for this Assigned Content Type.
             */
            archivedContentWebhook: function () {
                return _this.fetchLinkedResource('archived-content-webhook', {}, Webhook_1.Webhook);
            },
        };
        return _this;
    }
    return AssignedContentType;
}(HalResource_1.HalResource));
exports.AssignedContentType = AssignedContentType;
/**
 * @hidden
 */
var AssignedContentTypePage = /** @class */ (function (_super) {
    __extends(AssignedContentTypePage, _super);
    function AssignedContentTypePage(data) {
        return _super.call(this, 'assigned-content-types', AssignedContentType, data) || this;
    }
    return AssignedContentTypePage;
}(Page_1.Page));
exports.AssignedContentTypePage = AssignedContentTypePage;
