// dc-management-sdk-js/build/main/lib/model/ContentTypeSchema.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContentTypeSchemaPage = exports.ContentTypeSchema = exports.ValidationLevel = void 0;
var HalResource_1 = require("../hal/models/HalResource");
var Hub_1 = require("./Hub");
var Page_1 = require("./Page");
/**
 * Supported validation levels
 */
var ValidationLevel;
(function (ValidationLevel) {
    ValidationLevel["SLOT"] = "SLOT";
    ValidationLevel["CONTENT_TYPE"] = "CONTENT_TYPE";
    ValidationLevel["PARTIAL"] = "PARTIAL";
})(ValidationLevel = exports.ValidationLevel || (exports.ValidationLevel = {}));
/**
 * Content type schema
 */
var ContentTypeSchema = /** @class */ (function (_super) {
    __extends(ContentTypeSchema, _super);
    function ContentTypeSchema() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /**
         * Resources and actions related to a ContentTypeSchema
         */
        _this.related = {
            /**
             * Retrieves the Hub this content type schema is stored in
             */
            hub: function () { return _this.fetchLinkedResource('hub', {}, Hub_1.Hub); },
            /**
             * Updated content type schema
             * @param updated
             */
            update: function (mutation) {
                return _this.updateResource(mutation, ContentTypeSchema);
            },
            /**
             * Archive content type schema
             */
            archive: function () {
                return _this.performActionThatReturnsResource('archive', {}, { version: _this.version }, ContentTypeSchema);
            },
            /**
             * Unarchive content type schema
             */
            unarchive: function () {
                return _this.performActionThatReturnsResource('unarchive', {}, { version: _this.version }, ContentTypeSchema);
            },
        };
        return _this;
    }
    return ContentTypeSchema;
}(HalResource_1.HalResource));
exports.ContentTypeSchema = ContentTypeSchema;
/**
 * @hidden
 */
var ContentTypeSchemaPage = /** @class */ (function (_super) {
    __extends(ContentTypeSchemaPage, _super);
    function ContentTypeSchemaPage(data) {
        return _super.call(this, 'content-type-schemas', ContentTypeSchema, data) || this;
    }
    return ContentTypeSchemaPage;
}(Page_1.Page));
exports.ContentTypeSchemaPage = ContentTypeSchemaPage;
