// dc-management-sdk-js/build/main/lib/model/ContentItem.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContentItemsFacets = exports.ContentItemsPage = exports.ContentItem = exports.FacetedContentItem = void 0;
var HalResource_1 = require("../hal/models/HalResource");
var ContentRepository_1 = require("./ContentRepository");
var LocalizationJob_1 = require("./LocalizationJob");
var Page_1 = require("./Page");
var Facets_1 = require("./Facets");
var HttpRequest_1 = require("../http/HttpRequest");
var BaseContentItem = /** @class */ (function (_super) {
    __extends(BaseContentItem, _super);
    function BaseContentItem() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /**
         * Resources and actions related to a Content Item
         */
        _this.related = {
            /**
             * Retrieves a specific version of the content item
             * @param version Version number requested
             */
            contentItemVersion: function (version) {
                return _this.fetchLinkedResource('content-item-version', { version: version }, ContentItem);
            },
            /**
             * Retrieves the ContentRepository this content item is stored in
             */
            contentRepository: function () {
                return _this.fetchLinkedResource('content-repository', {}, ContentRepository_1.ContentRepository);
            },
            /**
             * Sets a locale of the form ll-CC (language, country code)
             * @param locale Locale code
             */
            setLocale: function (localeDefinition) {
                return _this.performActionThatReturnsResource('set-locale', {}, { locale: localeDefinition, version: _this.version }, ContentItem);
            },
            /**
             * Create localizations of the content item
             * @param locales Array of locales to create
             */
            localize: function (localesList) {
                return _this.performActionThatReturnsResource('create-localizations', {}, { locales: localesList, version: _this.version }, LocalizationJob_1.LocalizationJob);
            },
            /**
             * Get localizations of the content item
             */
            localizations: function (options) {
                return _this.fetchLinkedResource('localizations', options, ContentItemsPage);
            },
            /**
             * Updates this Content Item with the changes in the mutation parameter.
             * You must provide the current version number in the mutation
             * to avoid overwriting other user's changes.
             */
            update: function (mutation) {
                return _this.updateResource(mutation, ContentItem);
            },
            /**
             * Archive content item
             */
            archive: function () {
                return _this.performActionThatReturnsResource('archive', {}, { version: _this.version }, ContentItem);
            },
            /**
             * Unarchive content item
             */
            unarchive: function () {
                return _this.performActionThatReturnsResource('unarchive', {}, { version: _this.version }, ContentItem);
            },
            /**
             * Assign a WorkflowState
             */
            assignWorkflowState: function (workflowState) {
                return _this.performActionThatReturnsResource('edit-workflow', {}, { version: _this.version, state: workflowState.id }, ContentItem, HttpRequest_1.HttpMethod.PATCH);
            },
        };
        return _this;
    }
    return BaseContentItem;
}(HalResource_1.HalResource));
/**
 * Class representing a Faceted Content Item.
 * A faceted content item does not include a body, but includes the schema of the content item
 */
var FacetedContentItem = /** @class */ (function (_super) {
    __extends(FacetedContentItem, _super);
    function FacetedContentItem() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return FacetedContentItem;
}(BaseContentItem));
exports.FacetedContentItem = FacetedContentItem;
/**
 * Class representing the [Content Item](https://amplience.com/docs/api/dynamic-content/management/#tag/Content-Items) resource.
 * Content Items are instances of content created from a content type.
 */
var ContentItem = /** @class */ (function (_super) {
    __extends(ContentItem, _super);
    function ContentItem() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return ContentItem;
}(BaseContentItem));
exports.ContentItem = ContentItem;
/**
 * @hidden
 */
var ContentItemsPage = /** @class */ (function (_super) {
    __extends(ContentItemsPage, _super);
    function ContentItemsPage(data) {
        return _super.call(this, 'content-items', ContentItem, data) || this;
    }
    return ContentItemsPage;
}(Page_1.Page));
exports.ContentItemsPage = ContentItemsPage;
/**
 * @hidden
 */
var ContentItemsFacets = /** @class */ (function (_super) {
    __extends(ContentItemsFacets, _super);
    function ContentItemsFacets(data) {
        return _super.call(this, 'content-items', FacetedContentItem, data) || this;
    }
    return ContentItemsFacets;
}(Facets_1.FacetsResponse));
exports.ContentItemsFacets = ContentItemsFacets;
