// dc-management-sdk-js/build/main/lib/model/ContentType.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContentTypePage = exports.ContentType = void 0;
var HalResource_1 = require("../hal/models/HalResource");
var ContentTypeCachedSchema_1 = require("./ContentTypeCachedSchema");
var Page_1 = require("./Page");
/**
 * Class representing the [Content Type](https://amplience.com/docs/api/dynamic-content/management/#tag/Content-Types) resource.
 * Content types are JSON schemas that define a type of content to be created, including its structure, format and validation rules.
 */
var ContentType = /** @class */ (function (_super) {
    __extends(ContentType, _super);
    function ContentType() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /**
         * Resources and actions related to a ContentType
         */
        _this.related = {
            /**
             * Update a Content Type with the provided changes
             */
            update: function (mutation) {
                return _this.updateResource(mutation, ContentType);
            },
            /**
             * Archive content type
             */
            archive: function () {
                return _this.performActionThatReturnsResource('archive', {}, {}, ContentType);
            },
            /**
             * Unarchive content type
             */
            unarchive: function () {
                return _this.performActionThatReturnsResource('unarchive', {}, {}, ContentType);
            },
            contentTypeSchema: {
                /**
                 * Get the associated JSON schema document for a content type
                 */
                get: function () {
                    return _this.fetchLinkedResource('content-type-schema', {}, ContentTypeCachedSchema_1.ContentTypeCachedSchema);
                },
                /**
                 * Sync a content type with the associated JSON schema document
                 */
                update: function (mutation) {
                    if (mutation === void 0) { mutation = new ContentTypeCachedSchema_1.ContentTypeCachedSchema(); }
                    return _this.updateLinkedResource('content-type-schema', {}, mutation, ContentTypeCachedSchema_1.ContentTypeCachedSchema);
                },
            },
        };
        return _this;
    }
    return ContentType;
}(HalResource_1.HalResource));
exports.ContentType = ContentType;
/**
 * @hidden
 */
var ContentTypePage = /** @class */ (function (_super) {
    __extends(ContentTypePage, _super);
    function ContentTypePage(data) {
        return _super.call(this, 'content-types', ContentType, data) || this;
    }
    return ContentTypePage;
}(Page_1.Page));
exports.ContentTypePage = ContentTypePage;
