// dc-management-sdk-js/build/main/lib/model/Event.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventsPage = exports.Event = void 0;
var HalResource_1 = require("../hal/models/HalResource");
var Edition_1 = require("./Edition");
var Hub_1 = require("./Hub");
var Page_1 = require("./Page");
/**
 * Class representing the [Event](https://amplience.com/docs/api/dynamic-content/management/#tag/Events) resource.
 * An Event represents a major date in the calendar, such as Christmas or Black Friday. An Event may span multiple days and start and end at any given UTC time.
 */
var Event = /** @class */ (function (_super) {
    __extends(Event, _super);
    function Event() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /**
         * Resources and actions related to an Event
         */
        _this.related = {
            /**
             * Archive Event
             */
            archive: function () {
                return _this.performActionThatReturnsResource('archive', {}, {}, Event);
            },
            /**
             * Delete Event
             */
            delete: function () { return _this.deleteResource(); },
            /**
             * Retrieves the Hub this event is stored in
             */
            hub: function () { return _this.fetchLinkedResource('hub', {}, Hub_1.Hub); },
            editions: {
                /**
                 * Creates an Edition inside the Event
                 * @param resource The new Edition to create
                 */
                create: function (resource) {
                    return _this.createLinkedResource('create-edition', {}, resource, Edition_1.Edition);
                },
                /**
                 * Retrieves a list of Editions associated with this Event.
                 * @param options Pagination options
                 */
                list: function (options) {
                    return _this.fetchLinkedResource('editions', options, Edition_1.EditionsPage);
                },
            },
            /**
             * Updates this Event with the changes in the mutation parameter.
             * @param mutation Mutated event
             */
            update: function (mutation) {
                return _this.updateResource(mutation, Event);
            },
        };
        return _this;
    }
    return Event;
}(HalResource_1.HalResource));
exports.Event = Event;
/**
 * @hidden
 */
var EventsPage = /** @class */ (function (_super) {
    __extends(EventsPage, _super);
    function EventsPage(data) {
        return _super.call(this, 'events', Event, data) || this;
    }
    return EventsPage;
}(Page_1.Page));
exports.EventsPage = EventsPage;
