// dc-management-sdk-js/build/main/lib/model/Folder.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.FoldersPage = exports.Folder = void 0;
var HalResource_1 = require("../hal/models/HalResource");
var CURIEs_1 = require("../hal/services/CURIEs");
var ContentItem_1 = require("./ContentItem");
var ContentRepository_1 = require("./ContentRepository");
var Page_1 = require("./Page");
var Folder = /** @class */ (function (_super) {
    __extends(Folder, _super);
    function Folder() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /**
         * Resources and actions related to a Folder
         */
        _this.related = {
            /**
             * Retrieves the ContentRepository this folder is stored in
             */
            contentRepository: function () {
                return _this.fetchLinkedResource('content-repository', {}, ContentRepository_1.ContentRepository);
            },
            folders: {
                /**
                 * Retrieves the parent folder of this Folder
                 */
                parent: function () {
                    return _this.fetchLinkedResource('parent-folder', {}, Folder);
                },
                /**
                 * Retrieves the list of sub-folders contained within this Folder
                 * @param options Pagination options
                 */
                list: function (options) {
                    var path = CURIEs_1.CURIEs.expand("folders/" + _this.id + "/folders{?page,size,sort}", options);
                    return _this.client.fetchResource(path, FoldersPage);
                },
                /**
                 * Creates a folder beneath the folder
                 * @param resource The new Folder to create
                 */
                create: function (resource) {
                    return _this.createLinkedResource('create-folder', {}, resource, Folder);
                },
            },
            contentItems: {
                /**
                 * Retrieves a list of Content Items stored within this Folder
                 * @param options Pagination options
                 */
                list: function (options) {
                    return _this.fetchLinkedResource('content-items', options, ContentItem_1.ContentItemsPage);
                },
            },
        };
        return _this;
    }
    return Folder;
}(HalResource_1.HalResource));
exports.Folder = Folder;
/**
 * @hidden
 */
var FoldersPage = /** @class */ (function (_super) {
    __extends(FoldersPage, _super);
    function FoldersPage(data) {
        return _super.call(this, 'folders', Folder, data) || this;
    }
    return FoldersPage;
}(Page_1.Page));
exports.FoldersPage = FoldersPage;
