// dc-management-sdk-js/build/main/lib/model/HierarchyChildren.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.HierarchyChildren = void 0;
var HalResource_1 = require("../hal/models/HalResource");
var HierarchyNode_1 = require("./HierarchyNode");
var ContentItem_1 = require("./ContentItem");
var HierarchyChildren = /** @class */ (function (_super) {
    __extends(HierarchyChildren, _super);
    function HierarchyChildren(data) {
        var _this = _super.call(this, data) || this;
        _this.related = {
            contentItem: {
                /**
                 * Retrieve the content item associated with this node
                 */
                get: function () {
                    return _this.fetchLinkedResource('content-item', {}, ContentItem_1.ContentItem);
                },
            },
        };
        _this.children = _this.children.map(function (node) { return new HierarchyNode_1.HierarchyNode(node); });
        return _this;
    }
    HierarchyChildren.prototype.setClient = function (client) {
        this.client = client;
        this.children.forEach(function (child) {
            child.setClient(client);
        });
    };
    HierarchyChildren.prototype.toJson = function () {
        return this.toJSON();
    };
    HierarchyChildren.prototype.toJSON = function () {
        return Object.assign(_super.prototype.toJSON.call(this), {
            children: this.children.map(function (parent) { return parent.toJSON(); }),
        });
    };
    return HierarchyChildren;
}(HalResource_1.HalResource));
exports.HierarchyChildren = HierarchyChildren;
