// dc-management-sdk-js/build/main/lib/model/Hub.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HubsPage = exports.Hub = void 0;
var HalResource_1 = require("../hal/models/HalResource");
var ContentItem_1 = require("./ContentItem");
var ContentRepository_1 = require("./ContentRepository");
var ContentType_1 = require("./ContentType");
var ContentTypeSchema_1 = require("./ContentTypeSchema");
var Edition_1 = require("./Edition");
var Event_1 = require("./Event");
var Page_1 = require("./Page");
var SearchIndex_1 = require("./SearchIndex");
var Settings_1 = require("./Settings");
var Webhook_1 = require("./Webhook");
var WorkflowState_1 = require("./WorkflowState");
var Extension_1 = require("./Extension");
var SnapshotResultList_1 = require("./SnapshotResultList");
/**
 * Class representing the [Hub](https://amplience.com/docs/api/dynamic-content/management/#resources-hubs) resource.
 * Hubs are containers for multiple repositories including media, content, content types.
 */
var Hub = /** @class */ (function (_super) {
    __extends(Hub, _super);
    function Hub() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /**
         * Resources and actions related to a Hub
         */
        _this.related = {
            contentTypeSchema: {
                /**
                 * Creates a new Content Type Schema
                 * @param resource
                 */
                create: function (resource) {
                    return _this.createLinkedResource('create-content-type-schema', {}, resource, ContentTypeSchema_1.ContentTypeSchema);
                },
                /**
                 * Retrieves a list of Content Type Schemas associated with this Hub
                 * @param options Pagination options
                 */
                list: function (options) {
                    return _this.fetchLinkedResource('list-content-type-schemas', options, ContentTypeSchema_1.ContentTypeSchemaPage);
                },
            },
            contentRepositories: {
                /**
                 * Retrieves a list of Content Repositories associated with this Hub
                 * @param options Pagination options
                 */
                list: function (options) {
                    return _this.fetchLinkedResource('content-repositories', options, ContentRepository_1.ContentRepositoriesPage);
                },
                /**
                 * Find all the Content Repositories that contain a feature
                 * @param options Options
                 */
                findByFeature: function (feature, options) {
                    return _this.client.fetchLinkedResource({
                        href: "hubs/" + _this.id + "/content-repositories/search/findByFeaturesContaining?feature={feature}{&page,size,sort}",
                        templated: true,
                    }, __assign({ feature: feature }, options), ContentRepository_1.ContentRepositoriesPage);
                },
            },
            contentTypes: {
                /**
                 * Retrieves a list of Content Types associated with this Hub
                 * @param options Pagination options
                 */
                list: function (options) {
                    return _this.fetchLinkedResource('content-types', options, ContentType_1.ContentTypePage);
                },
                /**
                 * Register a Content Type on a hub so it can be enabled on a repository
                 */
                register: function (resource) {
                    return _this.createLinkedResource('register-content-type', {}, resource, ContentType_1.ContentType);
                },
                /**
                 * Get a content type by its id
                 */
                get: function (id) {
                    return _this.client.fetchResource("content-types/" + id, ContentType_1.ContentType);
                },
            },
            contentItems: {
                /**
                 * Facet content items and search by text. See [Text Search Query Syntax](https://amplience.com/docs/api/dynamic-content/management/#section/Text-Search-Query-Syntax) for a summary of the query syntax.
                 */
                facet: function (facetQuery, options) {
                    return _this.performActionThatReturnsResource('facet-content-items', options, facetQuery, ContentItem_1.ContentItemsFacets);
                },
            },
            editions: {
                /**
                 * Find editions by date associated with this Hub
                 * @param options FindByDate & Pageable & Sortable options
                 */
                findByDate: function (options, projection) {
                    return _this.fetchLinkedResource('search-editions', __assign({ projection: projection }, options), Edition_1.EditionsPage);
                },
            },
            events: {
                /**
                 * Creates an Event inside this Hub
                 * @param resource
                 */
                create: function (resource) {
                    return _this.createLinkedResource('create-event', {}, resource, Event_1.Event);
                },
                /**
                 * Retrieves a list of Events associated with this Hub
                 * @param options Pagination options
                 */
                list: function (options) {
                    return _this.fetchLinkedResource('events', options, Event_1.EventsPage);
                },
            },
            extensions: {
                /**
                 * Creates an Extension inside this Hub
                 * @param resource
                 */
                create: function (resource) {
                    return _this.createLinkedResource('create-extension', {}, resource, Extension_1.Extension);
                },
                /**
                 * Get an Extension inside this Hub by its name
                 */
                getByName: function (name) {
                    return _this.fetchLinkedResource('extension-by-name', { name: name }, Extension_1.Extension);
                },
                /**
                 * Retrieves a list of Extension associated with this Hub
                 * @param options Pagination options
                 */
                list: function (options) {
                    return _this.fetchLinkedResource('extensions', options, Extension_1.ExtensionsPage);
                },
            },
            searchIndexes: {
                /**
                 * Create a search index for a given hub
                 * @param resource
                 */
                create: function (resource) {
                    return _this.createLinkedResource('create-algolia-search-index', {}, resource, SearchIndex_1.SearchIndex);
                },
                /**
                 * Get a search index by its id
                 */
                get: function (id) {
                    return _this.client.fetchResource("algolia-search/" + _this.id + "/indexes/" + id, SearchIndex_1.SearchIndex);
                },
                /**
                 * Retrieves a list of search indexes associated with this Hub
                 */
                list: function (parentId, projection, options) {
                    return _this.fetchLinkedResource('algolia-search-indexes', __assign({ parentId: parentId, projection: projection }, options), SearchIndex_1.SearchIndexesPage);
                },
            },
            settings: {
                /**
                 * Updates this hub settings with the changes in the mutation parameter.
                 */
                update: function (mutation) {
                    return _this.updateLinkedResource('update-settings', {
                        method: 'PATCH',
                    }, mutation, Settings_1.Settings);
                },
            },
            snapshots: {
                /**
                 * Creates a Snapshot inside this Hub
                 * @param request
                 */
                create: function (request) {
                    return _this.performActionThatReturnsResource('batch-create-snapshots', {}, request, SnapshotResultList_1.SnapshotResultList);
                },
            },
            webhooks: {
                /**
                 * Creates a Webhook inside this Hub
                 * @param resource
                 */
                create: function (resource) {
                    return _this.createLinkedResource('create-webhook', {}, resource, Webhook_1.Webhook);
                },
                /**
                 * Get a Webhook inside this Hub by its id
                 */
                get: function (id) {
                    return _this.client.fetchResource("hubs/" + _this.id + "/webhooks/" + id, Webhook_1.Webhook);
                },
                /**
                 * Retrieves a list of Webhooks associated with this Hub
                 * @param options Pagination options
                 */
                list: function (options) {
                    return _this.fetchLinkedResource('webhooks', options, Webhook_1.WebhooksPage);
                },
            },
            workflowStates: {
                /**
                 * Creates a Workflow State inside this Hub
                 * @param resource
                 */
                create: function (resource) {
                    return _this.createLinkedResource('create-workflow-state', {}, resource, WorkflowState_1.WorkflowState);
                },
                /**
                 * Get a Webhook State insidee this Hub by its id
                 * @param id
                 */
                get: function (id) {
                    return _this.client.fetchResource("workflow-states/" + id, WorkflowState_1.WorkflowState);
                },
                /**
                 * Retrieves a list of Workflow States associated with this Hub
                 * @param options Pagination options
                 */
                list: function (options) {
                    return _this.fetchLinkedResource('workflow-states', options, WorkflowState_1.WorkflowStatesPage);
                },
            },
        };
        return _this;
    }
    return Hub;
}(HalResource_1.HalResource));
exports.Hub = Hub;
/**
 * @hidden
 */
var HubsPage = /** @class */ (function (_super) {
    __extends(HubsPage, _super);
    function HubsPage(data) {
        return _super.call(this, 'hubs', Hub, data) || this;
    }
    return HubsPage;
}(Page_1.Page));
exports.HubsPage = HubsPage;
