// dc-management-sdk-js/build/main/lib/model/PublishingStatus.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContentItemPublishingStatus = exports.PublishingStatus = void 0;
/**
 * Edition publishing status
 */
var PublishingStatus;
(function (PublishingStatus) {
    PublishingStatus["DRAFT"] = "DRAFT";
    PublishingStatus["SCHEDULING"] = "SCHEDULING";
    PublishingStatus["SCHEDULED"] = "SCHEDULED";
    PublishingStatus["PUBLISHED"] = "PUBLISHED";
    PublishingStatus["UNSCHEDULING"] = "UNSCHEDULING";
    PublishingStatus["PUBLISHING"] = "PUBLISHING";
})(PublishingStatus = exports.PublishingStatus || (exports.PublishingStatus = {}));
/**
 * publishing status of the content item
 *
 * NONE - hasn't been published before
 *
 * EARLY - changes have been made since last publish
 *
 * LATEST - no changes have been made since last publish
 */
var ContentItemPublishingStatus;
(function (ContentItemPublishingStatus) {
    ContentItemPublishingStatus["NONE"] = "NONE";
    ContentItemPublishingStatus["EARLT"] = "EARLY";
    ContentItemPublishingStatus["LATEST"] = "LATEST";
})(ContentItemPublishingStatus = exports.ContentItemPublishingStatus || (exports.ContentItemPublishingStatus = {}));
