// dc-management-sdk-js/build/main/lib/model/SearchIndex.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SearchIndexesPage = exports.SearchIndex = void 0;
var HalResource_1 = require("../hal/models/HalResource");
var AssignedContentType_1 = require("./AssignedContentType");
var Page_1 = require("./Page");
var SearchIndexKey_1 = require("./SearchIndexKey");
var SearchIndexSettings_1 = require("./SearchIndexSettings");
var SearchIndexStatistics_1 = require("./SearchIndexStatistics");
var SearchIndexTopHits_1 = require("./SearchIndexTopHits");
var SearchIndexTopSearches_1 = require("./SearchIndexTopSearches");
var SearchIndexSearchesWithNoResults_1 = require("./SearchIndexSearchesWithNoResults");
var SearchIndexTopFiltersNoResultSearch_1 = require("./SearchIndexTopFiltersNoResultSearch");
var SearchIndexUsersCount_1 = require("./SearchIndexUsersCount");
var SearchIndexSearchesCount_1 = require("./SearchIndexSearchesCount");
var SearchIndexNoResultsRate_1 = require("./SearchIndexNoResultsRate");
var Hub_1 = require("./Hub");
/**
 * Class representing an Algolia Search Index.
 */
var SearchIndex = /** @class */ (function (_super) {
    __extends(SearchIndex, _super);
    function SearchIndex() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /**
         * Resources and actions related to a Search Index.
         */
        _this.related = {
            /**
             * Retrieves the Hub this search index is stored in
             */
            hub: function () { return _this.fetchLinkedResource('hub', {}, Hub_1.Hub); },
            clear: function () {
                return _this.performActionThatReturnsResource('clear', {}, {}, SearchIndex);
            },
            delete: function () { return _this.deleteResource(); },
            update: function (resource) {
                return _this.updateResource(resource, SearchIndex);
            },
            assignedContentTypes: {
                create: function (resource) {
                    return _this.createLinkedResource('create-assigned-content-types', {}, resource, AssignedContentType_1.AssignedContentType);
                },
                get: function (id) {
                    return _this.client.fetchResource(
                    // tslint:disable-next-line:no-string-literal
                    _this._links['self'].href + "/assigned-content-types/" + id, AssignedContentType_1.AssignedContentType);
                },
                list: function (options) {
                    return _this.fetchLinkedResource('assigned-content-types', { options: options }, AssignedContentType_1.AssignedContentTypePage);
                },
            },
            indexObject: {
                delete: function (id) {
                    // tslint:disable-next-line:no-string-literal
                    return _this.client.deleteResource(_this._links['self'].href + "/objects/" + id);
                },
            },
            keys: {
                get: function () {
                    return _this.fetchLinkedResource('hub-search-key', {}, SearchIndexKey_1.SearchIndexKey);
                },
            },
            replicas: {
                list: function (projection, options) {
                    return _this.fetchLinkedResource('list-replicas', { projection: projection, options: options }, SearchIndexesPage);
                },
            },
            settings: {
                get: function () {
                    return _this.fetchLinkedResource('settings', {}, SearchIndexSettings_1.SearchIndexSettings);
                },
                update: function (resource, forwardToReplicas) {
                    return _this.updateLinkedResource('update-settings', { forwardToReplicas: forwardToReplicas }, resource, SearchIndexSettings_1.SearchIndexSettings);
                },
            },
            stats: {
                get: function (period) {
                    return _this.fetchLinkedResource('stats', { period: period }, SearchIndexStatistics_1.SearchIndexStatistics);
                },
            },
            'top-searches': {
                get: function (_a) {
                    var _b = _a.clickAnalytics, clickAnalytics = _b === void 0 ? false : _b, orderBy = _a.orderBy, direction = _a.direction, startDate = _a.startDate, endDate = _a.endDate, limit = _a.limit, offset = _a.offset, tags = _a.tags, includeReplicas = _a.includeReplicas;
                    return _this.fetchLinkedResource('top-searches', {
                        clickAnalytics: clickAnalytics,
                        direction: direction,
                        endDate: endDate,
                        limit: limit,
                        offset: offset,
                        orderBy: orderBy,
                        startDate: startDate,
                        tags: tags,
                        includeReplicas: includeReplicas,
                    }, SearchIndexTopSearches_1.SearchIndexTopSearchesCollection);
                },
            },
            'top-hits': {
                get: function (_a) {
                    var search = _a.search, startDate = _a.startDate, endDate = _a.endDate, limit = _a.limit, offset = _a.offset, tags = _a.tags, includeReplicas = _a.includeReplicas;
                    return _this.fetchLinkedResource('top-hits', {
                        endDate: endDate,
                        limit: limit,
                        offset: offset,
                        search: search,
                        startDate: startDate,
                        tags: tags,
                        includeReplicas: includeReplicas,
                    }, SearchIndexTopHits_1.SearchIndexTopHitsCollection);
                },
            },
            'searches-with-no-results': {
                get: function (_a) {
                    var startDate = _a.startDate, endDate = _a.endDate, limit = _a.limit, offset = _a.offset, tags = _a.tags, includeReplicas = _a.includeReplicas;
                    return _this.fetchLinkedResource('searches-with-no-results', {
                        endDate: endDate,
                        limit: limit,
                        offset: offset,
                        startDate: startDate,
                        tags: tags,
                        includeReplicas: includeReplicas,
                    }, SearchIndexSearchesWithNoResults_1.SearchIndexSearchesWithNoResultsCollection);
                },
            },
            'top-filters-no-result-search': {
                get: function (_a) {
                    var search = _a.search, startDate = _a.startDate, endDate = _a.endDate, limit = _a.limit, offset = _a.offset, tags = _a.tags, includeReplicas = _a.includeReplicas;
                    return _this.fetchLinkedResource('top-filters-no-result-search', {
                        endDate: endDate,
                        limit: limit,
                        offset: offset,
                        search: search,
                        startDate: startDate,
                        tags: tags,
                        includeReplicas: includeReplicas,
                    }, SearchIndexTopFiltersNoResultSearch_1.SearchIndexTopFiltersNoResultSearchCollection);
                },
            },
            'users-count': {
                get: function (_a) {
                    var startDate = _a.startDate, endDate = _a.endDate, tags = _a.tags, includeReplicas = _a.includeReplicas;
                    return _this.fetchLinkedResource('users-count', {
                        startDate: startDate,
                        endDate: endDate,
                        tags: tags,
                        includeReplicas: includeReplicas,
                    }, SearchIndexUsersCount_1.SearchIndexUsersCount);
                },
            },
            'searches-count': {
                get: function (_a) {
                    var startDate = _a.startDate, endDate = _a.endDate, tags = _a.tags, includeReplicas = _a.includeReplicas;
                    return _this.fetchLinkedResource('searches-count', {
                        startDate: startDate,
                        endDate: endDate,
                        tags: tags,
                        includeReplicas: includeReplicas,
                    }, SearchIndexSearchesCount_1.SearchIndexSearchesCount);
                },
            },
            'no-results-rate': {
                get: function (_a) {
                    var startDate = _a.startDate, endDate = _a.endDate, tags = _a.tags, includeReplicas = _a.includeReplicas;
                    return _this.fetchLinkedResource('no-results-rate', {
                        startDate: startDate,
                        endDate: endDate,
                        tags: tags,
                        includeReplicas: includeReplicas,
                    }, SearchIndexNoResultsRate_1.SearchIndexNoResultsRate);
                },
            },
        };
        return _this;
    }
    return SearchIndex;
}(HalResource_1.HalResource));
exports.SearchIndex = SearchIndex;
/**
 * @hidden
 */
var SearchIndexesPage = /** @class */ (function (_super) {
    __extends(SearchIndexesPage, _super);
    function SearchIndexesPage(data) {
        return _super.call(this, 'indexes', SearchIndex, data) || this;
    }
    return SearchIndexesPage;
}(Page_1.Page));
exports.SearchIndexesPage = SearchIndexesPage;
