// dc-management-sdk-js/build/main/lib/model/SearchIndexTopSearches.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SearchIndexTopSearchesCollection = exports.SearchIndexTopSearches = exports.SearchesOrderBy = void 0;
var HalResource_1 = require("../hal/models/HalResource");
var EmbeddedCollection_1 = require("./EmbeddedCollection");
var SearchesOrderBy;
(function (SearchesOrderBy) {
    SearchesOrderBy["SEARCH_COUNT"] = "searchCount";
    SearchesOrderBy["CLICK_THROUGH_RATE"] = "clickThroughRate";
    SearchesOrderBy["CONVERSION_RATE"] = "conversionRate";
    SearchesOrderBy["AVERAGE_CLICK_POSITION"] = "averageClickPosition";
})(SearchesOrderBy = exports.SearchesOrderBy || (exports.SearchesOrderBy = {}));
/**
 * Class representing search analytics for an Algolia Search Index
 */
var SearchIndexTopSearches = /** @class */ (function (_super) {
    __extends(SearchIndexTopSearches, _super);
    function SearchIndexTopSearches() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return SearchIndexTopSearches;
}(HalResource_1.HalResource));
exports.SearchIndexTopSearches = SearchIndexTopSearches;
/**
 * @hidden
 */
var SearchIndexTopSearchesCollection = /** @class */ (function (_super) {
    __extends(SearchIndexTopSearchesCollection, _super);
    function SearchIndexTopSearchesCollection(data) {
        return _super.call(this, 'top-searches', SearchIndexTopSearches, data) || this;
    }
    return SearchIndexTopSearchesCollection;
}(EmbeddedCollection_1.EmbeddedCollection));
exports.SearchIndexTopSearchesCollection = SearchIndexTopSearchesCollection;
