// dc-management-sdk-js/build/main/lib/model/Webhook.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebhooksPage = exports.Webhook = void 0;
var HalResource_1 = require("../hal/models/HalResource");
var Hub_1 = require("./Hub");
var Page_1 = require("./Page");
/**
 * Class representing the [Webhook](https://amplience.com/docs/api/dynamic-content/management/#tag/Webhooks) resource.
 * A Webhook is a HTTP callback: a HTTP request that occurs when something happens; a simple event-notification via the HTTP protocol.
 */
var Webhook = /** @class */ (function (_super) {
    __extends(Webhook, _super);
    function Webhook() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        /**
         * Resources and actions related to a Webhook
         */
        _this.related = {
            /**
             * Retrieves the Hub this webhook is stored in
             */
            hub: function () { return _this.fetchLinkedResource('hub', {}, Hub_1.Hub); },
            /**
             * Delete Webhook
             */
            delete: function () { return _this.deleteResource(); },
            /**
             * Updates this Webhook with the changes in the mutation parameter.
             * @param mutation
             */
            update: function (mutation) {
                return _this.updateResource(mutation, Webhook);
            },
        };
        return _this;
    }
    Webhook.prototype.parse = function (data) {
        _super.prototype.parse.call(this, data);
        if (this.headers) {
            this.headers.forEach(function (header) {
                // secret can come back as null, lets convert this to false
                if (header.secret == null) {
                    header.secret = false;
                }
            });
        }
    };
    return Webhook;
}(HalResource_1.HalResource));
exports.Webhook = Webhook;
/**
 * @hidden
 */
var WebhooksPage = /** @class */ (function (_super) {
    __extends(WebhooksPage, _super);
    function WebhooksPage(data) {
        return _super.call(this, 'webhooks', Webhook, data) || this;
    }
    return WebhooksPage;
}(Page_1.Page));
exports.WebhooksPage = WebhooksPage;
