// dc-management-sdk-js/build/main/lib/DynamicContent.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DynamicContent = void 0;
var HalClient_1 = require("./hal/services/HalClient");
var AxiosHttpClient_1 = require("./http/AxiosHttpClient");
var ContentItem_1 = require("./model/ContentItem");
var ContentRepository_1 = require("./model/ContentRepository");
var ContentType_1 = require("./model/ContentType");
var ContentTypeSchema_1 = require("./model/ContentTypeSchema");
var Edition_1 = require("./model/Edition");
var Event_1 = require("./model/Event");
var Folder_1 = require("./model/Folder");
var Hub_1 = require("./model/Hub");
var Snapshot_1 = require("./model/Snapshot");
var OAuth2Client_1 = require("./oauth2/services/OAuth2Client");
var HierarchyParents_1 = require("./model/HierarchyParents");
var HierarchyChildren_1 = require("./model/HierarchyChildren");
var WorkflowState_1 = require("./model/WorkflowState");
var Extension_1 = require("./model/Extension");
/**
 * Amplience Dynamic Content API client.
 *
 * Your application should avoid where possible creating a new client for each request.
 * Authentication tokens are cached by the client and only re-requested when
 * they expire, allowing performance to be improved by reusing the client across requests.
 *
 * If multiple sets of credentials are required your application should create one client
 * per credential set.
 *
 * Example:
 *
 * ```typescript
 * const client = new DynamicContent({
 *    client_id: process.env.CLIENT_ID,
 *    client_secret: process.env.CLIENT_SECRET
 * });
 *
 * const repository = await client.contentRepositories.get('<REPO-ID>');
 *
 * const contentItem = new ContentItem();
 * contentItem.label = 'Homepage Article';
 * contentItem.body = {
 *      _meta: {
 *          schema: "https://github.com/techiedarren/dc-examples/blob/master/content-types/blocks/text-block.json"
 *      },
 *      paragraphs: [
 *          "Example article text..."
 *      ]
 * };
 *
 * await repository.related.contentItems.create(contentItem);
 * ```
 */
var DynamicContent = /** @class */ (function () {
    /**
     * Creates a Dynamic Content API client instance. You must provide credentials that will
     * be used to authenticate with the API.
     *
     * @param clientCredentials Api credentials used to generate an authentication token
     * @param dcConfig Optional configuration settings for Dynamic Content
     * @param httpClient Optional request settings, can be used to provide proxy settings, add interceptors etc
     */
    function DynamicContent(clientCredentials, dcConfig, httpClient) {
        var _this = this;
        /**
         * Hub Resources
         */
        this.hubs = {
            /**
             * Retrieve a hub resource by id
             * @param id hub id, previously generated on creation
             */
            get: function (id) {
                return _this.client.fetchResource("/hubs/" + id, Hub_1.Hub);
            },
            /**
             * Retrieve a list of hub resources shared with your client credentials.
             * @param options Pagination options
             */
            list: function (options) {
                return _this.client.fetchLinkedResource({ href: '/hubs{?page,size}', templated: true }, options, Hub_1.HubsPage);
            },
        };
        this.contentTypeSchemas = {
            /**
             * Gets a content type schemas
             * @param id ContentTypeSchema id
             */
            get: function (id) {
                return _this.client.fetchResource("/content-type-schemas/" + id, ContentTypeSchema_1.ContentTypeSchema);
            },
            /**
             * Gets a content type schemas
             * @param id ContentTypeSchema id
             * @param version Version
             */
            getByVersion: function (id, version) {
                return _this.client.fetchResource("/content-type-schemas/" + id + "/" + version, ContentTypeSchema_1.ContentTypeSchema);
            },
        };
        /**
         * Content Repository Resources
         */
        this.contentRepositories = {
            /**
             * Retrieve a content repository resource by id
             * @param id content repository id, previously generated on creation
             */
            get: function (id) {
                return _this.client.fetchResource("/content-repositories/" + id, ContentRepository_1.ContentRepository);
            },
        };
        /**
         * Hierarchy Resources
         */
        this.hierarchies = {
            parents: {
                /**
                 * Retrieve parents of the associated content item
                 * @param id content item id accociated content item
                 */
                get: function (id) {
                    return _this.client.fetchResource("/hierarchy-node/" + id + "/parents", HierarchyParents_1.HierarchyParents);
                },
            },
            children: {
                /**
                 * Retrieve children of the associated content item
                 * @param id content item id accociated content item
                 */
                get: function (id) {
                    return _this.client.fetchResource("/hierarchy-node/" + id + "/children", HierarchyChildren_1.HierarchyChildren);
                },
            },
        };
        /**
         * Content Item Resources
         */
        this.contentItems = {
            /**
             * Retrieve a content item resource by id
             * @param id content item id, previously generated on creation
             */
            get: function (id) {
                return _this.client.fetchResource("/content-items/" + id, ContentItem_1.ContentItem);
            },
        };
        /**
         * Content Type Resources
         */
        this.contentTypes = {
            /**
             * Retrieve a content type resource by id
             * @param id content type id, previously generated on creation
             */
            get: function (id) {
                return _this.client.fetchResource("/content-types/" + id, ContentType_1.ContentType);
            },
        };
        this.folders = {
            /**
             * Retrieve a folder resource by id
             * @param id folder id, previously generated on creation
             */
            get: function (id) {
                return _this.client.fetchResource("/folders/" + id, Folder_1.Folder);
            },
        };
        /**
         * Snapshot Resources
         */
        this.snapshots = {
            /**
             * Retrieve a snapshot resource by id
             * @param id snapshot id, previously generated on creation
             */
            get: function (id) {
                return _this.client.fetchResource("/snapshots/" + id, Snapshot_1.Snapshot);
            },
        };
        /**
         * Event Resources
         */
        this.events = {
            /**
             * Retrieve an event resource by id
             * @param id event id, previously generated on creation
             */
            get: function (id) {
                return _this.client.fetchResource("/events/" + id, Event_1.Event);
            },
        };
        /**
         * Edition Resources
         */
        this.editions = {
            /**
             * Retrieve an edition resource by id
             * @param id edition id, previously generated on creation
             */
            get: function (id) {
                return _this.client.fetchResource("/editions/" + id, Edition_1.Edition);
            },
        };
        /**
         * Extension Resources
         */
        this.extensions = {
            /**
             * Retrieve an extension resource by id
             * @param id extension id, previously generated on creation
             */
            get: function (id) {
                return _this.client.fetchResource("/extensions/" + id, Extension_1.Extension);
            },
        };
        /**
         * Workflow States
         */
        this.workflowStates = {
            /**
             * Retrieve a Workflow State by id
             * @param id of Workflow State, previously generated on creation
             */
            get: function (id) {
                return _this.client.fetchResource("/workflow-states/" + id, WorkflowState_1.WorkflowState);
            },
        };
        dcConfig = dcConfig || {};
        dcConfig.apiUrl = dcConfig.apiUrl || 'https://api.amplience.net/v2/content';
        dcConfig.authUrl = dcConfig.authUrl || 'https://auth.amplience.net';
        var httpClientInstance;
        if (httpClient !== undefined && 'request' in httpClient) {
            httpClientInstance = httpClient;
        }
        else {
            httpClientInstance = new AxiosHttpClient_1.AxiosHttpClient(httpClient === undefined ? {} : httpClient);
        }
        var tokenClient = this.createTokenClient(dcConfig, clientCredentials, httpClientInstance);
        this.client = this.createResourceClient(dcConfig, tokenClient, httpClientInstance);
    }
    DynamicContent.prototype.createTokenClient = function (dcConfig, clientCredentials, httpClient) {
        return new OAuth2Client_1.OAuth2Client(clientCredentials, {
            authUrl: dcConfig.authUrl,
        }, httpClient);
    };
    DynamicContent.prototype.createResourceClient = function (dcConfig, tokenProvider, httpClient) {
        return new HalClient_1.DefaultHalClient(dcConfig.apiUrl, httpClient, tokenProvider);
    };
    return DynamicContent;
}());
exports.DynamicContent = DynamicContent;
