// dc-management-sdk-js/build/module/lib/oauth2/services/OAuth2Client.js
import { HttpMethod } from '../../http/HttpRequest';
import { combineURLs } from '../../utils/URL';
/**
 * @hidden
 */
export class OAuth2Client {
    constructor(clientCredentials, { authUrl = 'https://auth.amplience.net' }, httpClient) {
        this.authUrl = authUrl;
        this.clientCredentials = clientCredentials;
        this.httpClient = httpClient;
    }
    /**
     * Requests an authentication token that can be used
     * to make requests to the Dynamic Content api.
     * Tokens are reused until they expire.
     *
     * @returns {Promise<AccessToken>}
     */
    async getToken() {
        if (this.inFlight != null) {
            return this.inFlight;
        }
        if (this.token != null && this.tokenExpires > Date.now()) {
            return this.token;
        }
        const payload = 'grant_type=client_credentials' +
            '&client_id=' +
            encodeURIComponent(this.clientCredentials.client_id) +
            '&client_secret=' +
            encodeURIComponent(this.clientCredentials.client_secret);
        const request = this.httpClient.request({
            data: payload,
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded',
            },
            method: HttpMethod.POST,
            url: combineURLs(this.authUrl, '/oauth/token'),
        });
        this.inFlight = request.then((response) => {
            if (typeof response.data !== 'object') {
                throw new Error('Unexpected response format from /oauth/token endpoint');
            }
            this.token = response.data;
            this.tokenExpires = Date.now() + this.token.expires_in * 1000;
            this.inFlight = null;
            return this.token;
        });
        return this.inFlight;
    }
}
