// dc-delivery-sdk-js/build/main/lib/content/model/Edition.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Edition = void 0;
/**
 * Class representing the Edition resource with helper functions.
 * An Edition is the main way of scheduling content to be published on a specific date.
 */
var Edition = /** @class */ (function () {
    /**
     * Creates a new Edition instance.
     * @param data JSON representation of the Edition model
     */
    function Edition(data) {
        if (data) {
            Object.assign(this, data);
        }
    }
    /**
     * Date when the edition should begin
     */
    Edition.prototype.getStartDate = function () {
        return new Date(this.start);
    };
    /**
     * Date when the edition should end
     */
    Edition.prototype.getEndDate = function () {
        return new Date(this.end);
    };
    /**
     * Export to JSON
     */
    Edition.prototype.toJSON = function () {
        return {
            id: this.id,
            start: this.start,
            end: this.end,
        };
    };
    return Edition;
}());
exports.Edition = Edition;
