// dc-management-sdk-js/build/main/lib/hal/utils/HalMock.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HalMocks = exports.HalMockResource = void 0;
var CURIEs_1 = require("../services/CURIEs");
/**
 * @hidden
 */
var HalMockResource = /** @class */ (function () {
    function HalMockResource(resource, mocks) {
        this.resource = resource;
        this.mocks = mocks;
    }
    HalMockResource.prototype.nestedResource = function (linkName, args, resource) {
        var link = this.resource._links[linkName];
        var href = CURIEs_1.CURIEs.expand(link.href, args);
        this.mocks.resource(resource, href);
        return this;
    };
    HalMockResource.prototype.nestedCollection = function (linkName, args, type, values) {
        var link = this.resource._links[linkName];
        var href = CURIEs_1.CURIEs.expand(link.href, args);
        this.mocks.collection(href, type, values);
        return this;
    };
    HalMockResource.prototype.nestedCreateResource = function (linkName, args, resource, errorCode) {
        if (errorCode === void 0) { errorCode = 200; }
        var link = this.resource._links[linkName];
        var href = CURIEs_1.CURIEs.expand(link.href, args);
        this.mocks.createResource(href, resource, errorCode);
        return this;
    };
    HalMockResource.prototype.nestedPutResource = function (linkName, args, resource) {
        var link = this.resource._links[linkName];
        var href = CURIEs_1.CURIEs.expand(link.href, args);
        this.mocks.putResource(href, resource);
        return this;
    };
    HalMockResource.prototype.nestedUpdateResource = function (linkName, args, resource) {
        var link = this.resource._links[linkName];
        var href = CURIEs_1.CURIEs.expand(link.href, args);
        this.mocks.updateResource(href, resource);
        return this;
    };
    HalMockResource.prototype.nestedDelete = function (linkName, args) {
        var link = this.resource._links[linkName];
        var href = CURIEs_1.CURIEs.expand(link.href, args);
        this.mocks.deleteResource(href);
        return this;
    };
    return HalMockResource;
}());
exports.HalMockResource = HalMockResource;
/**
 * @hidden
 */
var HalMocks = /** @class */ (function () {
    function HalMocks(mockInstance) {
        this.mockInstance = mockInstance;
    }
    HalMocks.prototype.resource = function (resource, url) {
        if (!url) {
            url = resource._links.self.href;
        }
        this.mockInstance.onGet(url).reply(200, resource);
        if (resource._links.delete) {
            this.mockInstance.onDelete(resource._links.delete.href).reply(204);
        }
        return new HalMockResource(resource, this);
    };
    HalMocks.prototype.collection = function (url, type, values) {
        var _a;
        this.mockInstance.onGet(url).reply(200, {
            _embedded: (_a = {},
                _a[type] = values,
                _a),
        });
    };
    HalMocks.prototype.createResource = function (url, resource, errorCode) {
        if (errorCode === void 0) { errorCode = 200; }
        this.mockInstance.onPost(url).reply(errorCode, resource);
        return new HalMockResource(resource, this);
    };
    HalMocks.prototype.putResource = function (url, resource) {
        this.mockInstance.onPut(url).reply(200, resource);
        return new HalMockResource(resource, this);
    };
    HalMocks.prototype.updateResource = function (url, resource) {
        this.mockInstance.onPatch(url).reply(200, resource);
        return new HalMockResource(resource, this);
    };
    HalMocks.prototype.deleteResource = function (url) {
        this.mockInstance.onDelete(url).reply(200);
    };
    HalMocks.prototype.postResource = function (url) {
        this.mockInstance.onPost(url).reply(204);
    };
    return HalMocks;
}());
exports.HalMocks = HalMocks;
