// dc-management-sdk-js/build/module/lib/hal/services/HalClient.js
import { HttpError } from '../../http/HttpError';
import { HttpMethod } from '../../http/HttpRequest';
import { combineURLs } from '../../utils/URL';
import { CURIEs } from './CURIEs';
/**
 * @hidden
 */
export class DefaultHalClient {
    constructor(baseUrl, httpClient, tokenProvider) {
        this.baseUrl = baseUrl;
        this.httpClient = httpClient;
        this.tokenProvider = tokenProvider;
    }
    async fetchLinkedResource(link, params, resourceConstructor) {
        let href = link.href;
        if (link.templated) {
            href = CURIEs.expand(href, params);
        }
        return this.fetchResource(href, resourceConstructor);
    }
    async fetchResource(path, resourceConstructor) {
        const response = await this.invoke({
            method: HttpMethod.GET,
            url: path,
        });
        return this.parse(response.data, resourceConstructor);
    }
    async createLinkedResource(link, params, resource, resourceConstructor) {
        let href = link.href;
        if (link.templated) {
            href = CURIEs.expand(href, params);
        }
        return this.createResource(href, resource, resourceConstructor);
    }
    async createResource(path, resource, resourceConstructor) {
        const response = await this.invoke({
            data: this.serialize(resource),
            method: HttpMethod.POST,
            url: path,
        });
        return this.parse(response.data, resourceConstructor);
    }
    async updateResource(path, resource, resourceConstructor) {
        const response = await this.invoke({
            data: this.serialize(resource),
            method: HttpMethod.PATCH,
            url: path,
        });
        return this.parse(response.data, resourceConstructor);
    }
    async updateLinkedResource(link, params, resource, resourceConstructor) {
        let href = link.href;
        if (link.templated) {
            href = CURIEs.expand(href, params);
        }
        return this.updateResource(href, resource, resourceConstructor);
    }
    async deleteLinkedResource(link, params) {
        let href = link.href;
        if (link.templated) {
            href = CURIEs.expand(href, params);
        }
        return this.deleteResource(href);
    }
    async deleteResource(path) {
        await this.invoke({
            method: HttpMethod.DELETE,
            url: path,
        });
        return Promise.resolve();
    }
    async performActionWithoutResourceResponse(link, params, data, method = HttpMethod.POST) {
        let href = link.href;
        if (link.templated) {
            href = CURIEs.expand(href, params);
        }
        await this.invoke({
            data: this.serialize(data),
            method,
            url: href,
        });
        return Promise.resolve();
    }
    async performActionThatReturnsResource(link, params, data, resourceConstructor, method = HttpMethod.POST) {
        let href = link.href;
        if (link.templated) {
            href = CURIEs.expand(href, params);
        }
        const response = await this.invoke({
            data: this.serialize(data),
            method,
            url: href,
        });
        return this.parse(response.data, resourceConstructor);
    }
    parse(data, resourceConstructor) {
        const instance = new resourceConstructor(data);
        instance.setClient(this);
        return instance;
    }
    serialize(data) {
        return JSON.parse(JSON.stringify(data));
    }
    async invoke(request) {
        const token = await this.tokenProvider.getToken();
        const fullRequest = {
            data: request.data,
            headers: {
                Authorization: 'bearer ' + token.access_token,
            },
            method: request.method,
            url: combineURLs(this.baseUrl, request.url),
        };
        return this.httpClient.request(fullRequest).then((response) => {
            if (response.status === 204) {
                return response;
            }
            if (response.status >= 200 && response.status < 300) {
                if (typeof response.data === 'string') {
                    response.data = JSON.parse(response.data);
                }
                return response;
            }
            else {
                throw new HttpError(`Request failed with status code ${response.status}: ${JSON.stringify(response.data)}`, fullRequest, response);
            }
        });
    }
}
