// dc-delivery-sdk-js/build/module/lib/content/coordinators/FilterBy.js
import { FilterByImpl } from './FilterByImpl';
export class FilterBy {
    constructor(config, contentClient) {
        this.config = config;
        this.contentClient = contentClient;
        this.requestConfig = {
            filterBy: [],
        };
        this.filterByService = new FilterByImpl(this.config, this.contentClient);
    }
    /**
     *  This function will help construct requests for filtering Content Items or Slots
     *
     * @param path - json path to the property you wish to filter by e.g `/_meta/schema`
     * @param value - value you want to return matches for
     *
     * @returns `FilterBy<Body>`
     */
    filterBy(path, value) {
        this.requestConfig.filterBy.push({
            path,
            value,
        });
        return this;
    }
    /**
     *
     *  equivalent to:
     *
     * ```ts
     *  client.filterBy('/_meta/schema', contentTypeUri)
     * ```
     *
     * @param contentTypeUri - Content Type Uri you want to filter
     *
     * @returns `FilterBy<Body>`
     */
    filterByContentType(value) {
        return this.filterBy(FilterBy.SCHEMA_PATH, value);
    }
    /**
     * Fetch content by parent id
     *
     * equivalent to:
     *
     * ```ts
     *  client.filterBy('/_meta/hierarchy/parentId', id)
     * ```
     *
     * @param id - ID of a Hierarchy Content Item
     *
     * @returns `FilterBy<Body>`
     */
    filterByParentId(value) {
        return this.filterBy(FilterBy.PARENT_PATH, value);
    }
    /**
     * Set sortable key based on schema [`trait:sortable`](https://amplience.com/docs/development/contentdelivery/filterandsort.html#sorttrait)
     *
     *
     * @param key - key you wish to sort by
     * @param order - order enum ASC, DESC
     *
     * @returns `FilterBy<Body>`
     */
    sortBy(key, order) {
        this.requestConfig.sortBy = {
            key,
            order,
        };
        return this;
    }
    page(size, cursor) {
        if (!this.requestConfig.page) {
            this.requestConfig.page = {};
        }
        if (arguments.length === 1 && typeof size === 'string') {
            this.requestConfig.page.cursor = size;
            return this;
        }
        if (typeof size === 'number') {
            this.requestConfig.page.size = size;
        }
        if (cursor) {
            this.requestConfig.page.cursor = cursor;
        }
        return this;
    }
    /**
     *
     * @param parameters - api options for response format
     *
     * @returns `FilterBy<Body>`
     */
    async request(parameters) {
        if (parameters) {
            this.requestConfig.parameters = parameters;
        }
        return this.filterByService.fetch(this.requestConfig);
    }
}
FilterBy.SCHEMA_PATH = '/_meta/schema';
FilterBy.PARENT_PATH = '/_meta/hierarchy/parentId';
