// dc-delivery-sdk-js/build/module/lib/content/coordinators/FilterByImpl.js
import { HttpError } from '../model/HttpError';
export class FilterByImpl {
    constructor(config, contentClient) {
        this.config = config;
        this.contentClient = contentClient;
    }
    async fetch(requestConfig) {
        try {
            if (!requestConfig.parameters?.locale && this.config?.locale) {
                requestConfig.parameters = Object.assign({}, {
                    ...(requestConfig.parameters || {}),
                    locale: this.config.locale,
                });
            }
            const { data } = await this.contentClient.post('content/filter', requestConfig);
            if (data.page.nextCursor) {
                const request = Object.assign({}, {
                    ...requestConfig,
                    page: {
                        ...(requestConfig.page || {}),
                        cursor: data.page.nextCursor,
                    },
                });
                data.page.next = () => this.fetch(request);
            }
            return data;
        }
        catch (err) {
            if (err.response) {
                throw new HttpError(err.response.status, err.response.data);
            }
            throw err;
        }
    }
}
