// dc-delivery-sdk-js/build/module/lib/content/coordinators/GetContentItemsV2Impl.js
import { HttpError } from '../model/HttpError';
/**
 * Utility that maps an array of ids or keys into the expected format of a `content/fetch` request body
 * @param property id or key
 * @param values array of ids or keys
 * @hidden
 */
function mapToRequests(property, values) {
    if (!Array.isArray(values)) {
        throw new TypeError('Expecting an array');
    }
    return values.map((value) => ({ [property]: String(value) }));
}
/**
 * Wraps `content/fetch`
 * @hidden
 */
export class GetContentItemsV2Impl {
    constructor(config, contentClient) {
        this.config = config;
        this.contentClient = contentClient;
    }
    async getContentItemsByKey(keys) {
        return this.getContentItems(mapToRequests('key', keys));
    }
    async getContentItemsById(ids) {
        return this.getContentItems(mapToRequests('id', ids));
    }
    async getContentItems(requests, parameters) {
        return this.fetchContentItems({
            parameters,
            requests,
        });
    }
    async fetchContentItems({ requests, parameters, }) {
        const defaultParameters = {
            depth: 'all',
            format: 'inlined',
        };
        if (this.config.locale) {
            defaultParameters.locale = this.config.locale;
        }
        const mergedParameters = {
            ...defaultParameters,
            ...(parameters || {}),
        };
        try {
            const { data } = await this.contentClient.post('content/fetch', {
                parameters: mergedParameters,
                requests,
            });
            return data;
        }
        catch (err) {
            if (err.response) {
                throw new HttpError(err.response.status, err.response.data);
            }
            throw err;
        }
    }
}
