// dc-delivery-sdk-js/build/module/lib/content/coordinators/GetContentItemV2Impl.js
import { ContentItem } from '../model/ContentItem';
import { encodeQueryString } from '../../utils/Url';
import { HttpError } from '../model/HttpError';
import { ContentNotFoundError } from '../model/ContentNotFoundError';
/**
 * @hidden
 */
export class GetContentItemV2Impl {
    constructor(config, contentClient, mapper) {
        this.config = config;
        this.contentClient = contentClient;
        this.mapper = mapper;
    }
    getContentItemByKey(key) {
        return this.getContentItem('key', key);
    }
    async getContentItemById(id) {
        return this.getContentItem('id', id);
    }
    async getContentItem(requestType, value) {
        const args = [
            ['depth', 'all'],
            ['format', 'inlined'],
        ];
        if (this.config.locale) {
            args.push(['locale', this.config.locale]);
        }
        const url = `content/${requestType}/${value}?${encodeQueryString(args)}`;
        let response;
        try {
            response = await this.contentClient.get(url);
        }
        catch (err) {
            if (err.response) {
                if (err.response.status === 404) {
                    throw new ContentNotFoundError(value);
                }
                else {
                    throw new HttpError(err.response.status, err.response.data);
                }
            }
            throw err;
        }
        const contentItem = new ContentItem();
        contentItem.body = this.mapper.toMappedContent(response.data.content);
        return contentItem;
    }
}
