// dc-delivery-sdk-js/build/main/lib/content/mapper/ContentMapper.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContentMapper = void 0;
var JsonWalker_1 = require("../../utils/JsonWalker");
var FragmentMeta_1 = require("../model/FragmentMeta");
var ContentMeta_1 = require("../model/ContentMeta");
var Image_1 = require("../../media/Image");
var Video_1 = require("../../media/Video");
var ContentReference_1 = require("../model/ContentReference");
/**
 * @hidden
 * ContentMapper provides functionality to convert simple JSON content into hydrated model classes
 * with helper functions. If a mapper is not defined for a schema it will remain as plain JSON.
 */
var ContentMapper = /** @class */ (function () {
    function ContentMapper(config) {
        this.config = config;
        this.mappers = [];
        this.registerBuiltInMappers();
    }
    /**
     * Registers a custom model / fn to hydrate a specific schema
     * @param schema JSON schema ID or Regex that matches against the JSON schema ID
     * @param fn Model Class or Function that can convert the schema
     */
    ContentMapper.prototype.addSchema = function (schema, fn) {
        if (typeof schema === 'string') {
            this.addCustomMapper(function (fragment) {
                if (fragment._meta.schema === schema) {
                    return fn(fragment);
                }
            });
        }
        else {
            this.addCustomMapper(function (fragment) {
                if (schema.test(fragment._meta.schema)) {
                    return fn(fragment);
                }
            });
        }
    };
    /**
     * Registers a custom function to hydrate content fragments
     * @param fn Mapper function
     */
    ContentMapper.prototype.addCustomMapper = function (fn) {
        this.mappers.push(fn);
    };
    /**
     * Converts the provided content into hydrated model classes
     * @param content Content to convert
     */
    ContentMapper.prototype.toMappedContent = function (content) {
        var _this = this;
        return JsonWalker_1.walkAndReplace(content, {
            afterWalk: function (node) {
                if (FragmentMeta_1.FragmentMeta.isFragment(node)) {
                    return _this.mapFragment(node);
                }
                return node;
            },
        });
    };
    /**
     * Converts a single fragment using the registered mappers
     * @param fragment Fragment to convert
     */
    ContentMapper.prototype.mapFragment = function (fragment) {
        for (var _i = 0, _a = this.mappers; _i < _a.length; _i++) {
            var mapper = _a[_i];
            var result = mapper(fragment);
            if (result) {
                return result;
            }
        }
        return fragment;
    };
    /**
     * Registers built in mappers
     */
    ContentMapper.prototype.registerBuiltInMappers = function () {
        this.addCustomMapper(this.convertContentReference.bind(this));
        this.addCustomMapper(this.convertContentMeta.bind(this));
        this.addCustomMapper(this.convertImage.bind(this));
        this.addCustomMapper(this.convertVideo.bind(this));
    };
    /**
     * Converts _meta inside Content Items into a ContentMeta instance
     * @param fragment
     */
    ContentMapper.prototype.convertContentMeta = function (fragment) {
        if (ContentMeta_1.ContentMeta.isContentBody(fragment)) {
            return __assign(__assign({}, fragment), { _meta: new ContentMeta_1.ContentMeta(fragment._meta) });
        }
    };
    /**
     * Converts content-link into an Content reference class instance
     * @param fragment
     */
    ContentMapper.prototype.convertContentReference = function (fragment) {
        if (ContentReference_1.ContentReference.isContentReference(fragment)) {
            var result = new ContentReference_1.ContentReference(fragment);
            return result;
        }
    };
    /**
     * Converts image-link into an Image class instance
     * @param fragment
     */
    ContentMapper.prototype.convertImage = function (fragment) {
        if (Image_1.Image.isImage(fragment)) {
            var result = new Image_1.Image(fragment, this.config);
            return result;
        }
    };
    /**
     * Converts image-link into an Image class instance
     * @param fragment
     */
    ContentMapper.prototype.convertVideo = function (fragment) {
        if (Video_1.Video.isVideo(fragment)) {
            var result = new Video_1.Video(fragment, this.config);
            return result;
        }
    };
    return ContentMapper;
}());
exports.ContentMapper = ContentMapper;
