// dc-management-sdk-js/build/module/lib/DynamicContent.js
import { DefaultHalClient } from './hal/services/HalClient';
import { AxiosHttpClient } from './http/AxiosHttpClient';
import { ContentItem } from './model/ContentItem';
import { ContentRepository } from './model/ContentRepository';
import { ContentType } from './model/ContentType';
import { ContentTypeSchema } from './model/ContentTypeSchema';
import { Edition } from './model/Edition';
import { Event } from './model/Event';
import { Folder } from './model/Folder';
import { Hub, HubsPage } from './model/Hub';
import { Snapshot } from './model/Snapshot';
import { OAuth2Client } from './oauth2/services/OAuth2Client';
import { HierarchyParents } from './model/HierarchyParents';
import { HierarchyChildren } from './model/HierarchyChildren';
import { WorkflowState } from './model/WorkflowState';
import { Extension } from './model/Extension';
/**
 * Amplience Dynamic Content API client.
 *
 * Your application should avoid where possible creating a new client for each request.
 * Authentication tokens are cached by the client and only re-requested when
 * they expire, allowing performance to be improved by reusing the client across requests.
 *
 * If multiple sets of credentials are required your application should create one client
 * per credential set.
 *
 * Example:
 *
 * ```typescript
 * const client = new DynamicContent({
 *    client_id: process.env.CLIENT_ID,
 *    client_secret: process.env.CLIENT_SECRET
 * });
 *
 * const repository = await client.contentRepositories.get('<REPO-ID>');
 *
 * const contentItem = new ContentItem();
 * contentItem.label = 'Homepage Article';
 * contentItem.body = {
 *      _meta: {
 *          schema: "https://github.com/techiedarren/dc-examples/blob/master/content-types/blocks/text-block.json"
 *      },
 *      paragraphs: [
 *          "Example article text..."
 *      ]
 * };
 *
 * await repository.related.contentItems.create(contentItem);
 * ```
 */
export class DynamicContent {
    /**
     * Creates a Dynamic Content API client instance. You must provide credentials that will
     * be used to authenticate with the API.
     *
     * @param clientCredentials Api credentials used to generate an authentication token
     * @param dcConfig Optional configuration settings for Dynamic Content
     * @param httpClient Optional request settings, can be used to provide proxy settings, add interceptors etc
     */
    constructor(clientCredentials, dcConfig, httpClient) {
        /**
         * Hub Resources
         */
        this.hubs = {
            /**
             * Retrieve a hub resource by id
             * @param id hub id, previously generated on creation
             */
            get: (id) => this.client.fetchResource(`/hubs/${id}`, Hub),
            /**
             * Retrieve a list of hub resources shared with your client credentials.
             * @param options Pagination options
             */
            list: (options) => this.client.fetchLinkedResource({ href: '/hubs{?page,size}', templated: true }, options, HubsPage),
        };
        this.contentTypeSchemas = {
            /**
             * Gets a content type schemas
             * @param id ContentTypeSchema id
             */
            get: (id) => this.client.fetchResource(`/content-type-schemas/${id}`, ContentTypeSchema),
            /**
             * Gets a content type schemas
             * @param id ContentTypeSchema id
             * @param version Version
             */
            getByVersion: (id, version) => this.client.fetchResource(`/content-type-schemas/${id}/${version}`, ContentTypeSchema),
        };
        /**
         * Content Repository Resources
         */
        this.contentRepositories = {
            /**
             * Retrieve a content repository resource by id
             * @param id content repository id, previously generated on creation
             */
            get: (id) => this.client.fetchResource(`/content-repositories/${id}`, ContentRepository),
        };
        /**
         * Hierarchy Resources
         */
        this.hierarchies = {
            parents: {
                /**
                 * Retrieve parents of the associated content item
                 * @param id content item id accociated content item
                 */
                get: (id) => this.client.fetchResource(`/hierarchy-node/${id}/parents`, HierarchyParents),
            },
            children: {
                /**
                 * Retrieve children of the associated content item
                 * @param id content item id accociated content item
                 */
                get: (id) => this.client.fetchResource(`/hierarchy-node/${id}/children`, HierarchyChildren),
            },
        };
        /**
         * Content Item Resources
         */
        this.contentItems = {
            /**
             * Retrieve a content item resource by id
             * @param id content item id, previously generated on creation
             */
            get: (id) => this.client.fetchResource(`/content-items/${id}`, ContentItem),
        };
        /**
         * Content Type Resources
         */
        this.contentTypes = {
            /**
             * Retrieve a content type resource by id
             * @param id content type id, previously generated on creation
             */
            get: (id) => this.client.fetchResource(`/content-types/${id}`, ContentType),
        };
        this.folders = {
            /**
             * Retrieve a folder resource by id
             * @param id folder id, previously generated on creation
             */
            get: (id) => this.client.fetchResource(`/folders/${id}`, Folder),
        };
        /**
         * Snapshot Resources
         */
        this.snapshots = {
            /**
             * Retrieve a snapshot resource by id
             * @param id snapshot id, previously generated on creation
             */
            get: (id) => this.client.fetchResource(`/snapshots/${id}`, Snapshot),
        };
        /**
         * Event Resources
         */
        this.events = {
            /**
             * Retrieve an event resource by id
             * @param id event id, previously generated on creation
             */
            get: (id) => this.client.fetchResource(`/events/${id}`, Event),
        };
        /**
         * Edition Resources
         */
        this.editions = {
            /**
             * Retrieve an edition resource by id
             * @param id edition id, previously generated on creation
             */
            get: (id) => this.client.fetchResource(`/editions/${id}`, Edition),
        };
        /**
         * Extension Resources
         */
        this.extensions = {
            /**
             * Retrieve an extension resource by id
             * @param id extension id, previously generated on creation
             */
            get: (id) => this.client.fetchResource(`/extensions/${id}`, Extension),
        };
        /**
         * Workflow States
         */
        this.workflowStates = {
            /**
             * Retrieve a Workflow State by id
             * @param id of Workflow State, previously generated on creation
             */
            get: (id) => this.client.fetchResource(`/workflow-states/${id}`, WorkflowState),
        };
        dcConfig = dcConfig || {};
        dcConfig.apiUrl = dcConfig.apiUrl || 'https://api.amplience.net/v2/content';
        dcConfig.authUrl = dcConfig.authUrl || 'https://auth.amplience.net';
        let httpClientInstance;
        if (httpClient !== undefined && 'request' in httpClient) {
            httpClientInstance = httpClient;
        }
        else {
            httpClientInstance = new AxiosHttpClient(httpClient === undefined ? {} : httpClient);
        }
        const tokenClient = this.createTokenClient(dcConfig, clientCredentials, httpClientInstance);
        this.client = this.createResourceClient(dcConfig, tokenClient, httpClientInstance);
    }
    createTokenClient(dcConfig, clientCredentials, httpClient) {
        return new OAuth2Client(clientCredentials, {
            authUrl: dcConfig.authUrl,
        }, httpClient);
    }
    createResourceClient(dcConfig, tokenProvider, httpClient) {
        return new DefaultHalClient(dcConfig.apiUrl, httpClient, tokenProvider);
    }
}
