// dc-management-sdk-js/build/module/lib/model/AssignedContentType.js
import { HalResource } from '../hal/models/HalResource';
import { Page } from './Page';
import { Webhook } from './Webhook';
/**
 * Class representing an Assigned Content Type for an Algolia Search Index.
 */
export class AssignedContentType extends HalResource {
    constructor() {
        super(...arguments);
        /**
         * Resources and actions related to an Assigned Content Type.
         */
        this.related = {
            /**
             * Unassigns the content type from the search index.
             */
            unassign: (id) => this.deleteLinkedResource('unassign', {
                id,
            }),
            /**
             * Recreates the webhooks for this Assigned Content Type.
             */
            recreateWebhook: () => this.performActionWithoutResourceResponse('recreate-webhook', {}, {}),
            /**
             * Gets the primary webhook for this Assigned Content Type.
             */
            webhook: () => this.fetchLinkedResource('webhook', {}, Webhook),
            /**
             * Gets the active content webhook for this Assigned Content Type.
             */
            activeContentWebhook: () => this.fetchLinkedResource('active-content-webhook', {}, Webhook),
            /**
             * Gets the archived content webhook for this Assigned Content Type.
             */
            archivedContentWebhook: () => this.fetchLinkedResource('archived-content-webhook', {}, Webhook),
        };
    }
}
/**
 * @hidden
 */
export class AssignedContentTypePage extends Page {
    constructor(data) {
        super('assigned-content-types', AssignedContentType, data);
    }
}
