// dc-management-sdk-js/build/module/lib/model/ContentType.js
import { HalResource } from '../hal/models/HalResource';
import { ContentTypeCachedSchema } from './ContentTypeCachedSchema';
import { Page } from './Page';
/**
 * Class representing the [Content Type](https://amplience.com/docs/api/dynamic-content/management/#tag/Content-Types) resource.
 * Content types are JSON schemas that define a type of content to be created, including its structure, format and validation rules.
 */
export class ContentType extends HalResource {
    constructor() {
        super(...arguments);
        /**
         * Resources and actions related to a ContentType
         */
        this.related = {
            /**
             * Update a Content Type with the provided changes
             */
            update: (mutation) => this.updateResource(mutation, ContentType),
            /**
             * Archive content type
             */
            archive: () => this.performActionThatReturnsResource('archive', {}, {}, ContentType),
            /**
             * Unarchive content type
             */
            unarchive: () => this.performActionThatReturnsResource('unarchive', {}, {}, ContentType),
            contentTypeSchema: {
                /**
                 * Get the associated JSON schema document for a content type
                 */
                get: () => this.fetchLinkedResource('content-type-schema', {}, ContentTypeCachedSchema),
                /**
                 * Sync a content type with the associated JSON schema document
                 */
                update: (mutation = new ContentTypeCachedSchema()) => this.updateLinkedResource('content-type-schema', {}, mutation, ContentTypeCachedSchema),
            },
        };
    }
}
/**
 * @hidden
 */
export class ContentTypePage extends Page {
    constructor(data) {
        super('content-types', ContentType, data);
    }
}
