// dc-management-sdk-js/build/module/lib/model/ContentTypeSchema.js
import { HalResource } from '../hal/models/HalResource';
import { Hub } from './Hub';
import { Page } from './Page';
/**
 * Supported validation levels
 */
export var ValidationLevel;
(function (ValidationLevel) {
    ValidationLevel["SLOT"] = "SLOT";
    ValidationLevel["CONTENT_TYPE"] = "CONTENT_TYPE";
    ValidationLevel["PARTIAL"] = "PARTIAL";
})(ValidationLevel || (ValidationLevel = {}));
/**
 * Content type schema
 */
export class ContentTypeSchema extends HalResource {
    constructor() {
        super(...arguments);
        /**
         * Resources and actions related to a ContentTypeSchema
         */
        this.related = {
            /**
             * Retrieves the Hub this content type schema is stored in
             */
            hub: () => this.fetchLinkedResource('hub', {}, Hub),
            /**
             * Updated content type schema
             * @param updated
             */
            update: (mutation) => this.updateResource(mutation, ContentTypeSchema),
            /**
             * Archive content type schema
             */
            archive: () => this.performActionThatReturnsResource('archive', {}, { version: this.version }, ContentTypeSchema),
            /**
             * Unarchive content type schema
             */
            unarchive: () => this.performActionThatReturnsResource('unarchive', {}, { version: this.version }, ContentTypeSchema),
        };
    }
}
/**
 * @hidden
 */
export class ContentTypeSchemaPage extends Page {
    constructor(data) {
        super('content-type-schemas', ContentTypeSchema, data);
    }
}
