// dc-management-sdk-js/build/module/lib/model/ContentRepository.js
import { HalResource } from '../hal/models/HalResource';
import { ContentItem, ContentItemsPage } from './ContentItem';
import { Folder, FoldersPage } from './Folder';
import { Hub } from './Hub';
import { Page } from './Page';
/**
 * Class representing the [Content Repository](https://amplience.com/docs/api/dynamic-content/management/#tag/Content-Repositories) resource.
 * Content repositories are containers where content can be stored.
 */
export class ContentRepository extends HalResource {
    constructor() {
        super(...arguments);
        /**
         * Resources and actions related to a Content Repository
         */
        this.related = {
            /**
             * Retrieves the Hub this content repository is stored in
             */
            hub: () => this.fetchLinkedResource('hub', {}, Hub),
            folders: {
                /**
                 * Retrieves the list of top-level folders within this Content Repository
                 */
                list: (options) => this.fetchLinkedResource('folders', options, FoldersPage),
                /**
                 * Creates a folder in the Content Repository
                 * @param resource The new Folder to create
                 */
                create: (resource) => this.createLinkedResource('create-folder', {}, resource, Folder),
            },
            contentItems: {
                /**
                 * Creates a content item inside the repository
                 * @param resource The new content item to create
                 */
                create: (resource) => this.createLinkedResource('create-content-item', {}, resource, ContentItem),
                /**
                 * Retrieves a list of Content Items stored within this Content Repository
                 */
                list: (options) => this.fetchLinkedResource('content-items', options, ContentItemsPage),
            },
            contentTypes: {
                /**
                 * Assigns a content type to the repository
                 */
                assign: (contentTypeId) => this.performActionThatReturnsResource('assign-content-type', {}, { contentTypeId }, ContentRepository),
                /**
                 * Unassign a content type from this repository
                 */
                unassign: (contentTypeId) => this.deleteLinkedResource('unassign-content-type', {
                    id: contentTypeId,
                }),
            },
        };
    }
}
/**
 * @hidden
 */
export class ContentRepositoriesPage extends Page {
    constructor(data) {
        super('content-repositories', ContentRepository, data);
    }
}
