// dc-management-sdk-js/build/module/lib/model/Event.js
import { HalResource } from '../hal/models/HalResource';
import { Edition, EditionsPage } from './Edition';
import { Hub } from './Hub';
import { Page } from './Page';
/**
 * Class representing the [Event](https://amplience.com/docs/api/dynamic-content/management/#tag/Events) resource.
 * An Event represents a major date in the calendar, such as Christmas or Black Friday. An Event may span multiple days and start and end at any given UTC time.
 */
export class Event extends HalResource {
    constructor() {
        super(...arguments);
        /**
         * Resources and actions related to an Event
         */
        this.related = {
            /**
             * Archive Event
             */
            archive: () => this.performActionThatReturnsResource('archive', {}, {}, Event),
            /**
             * Delete Event
             */
            delete: () => this.deleteResource(),
            /**
             * Retrieves the Hub this event is stored in
             */
            hub: () => this.fetchLinkedResource('hub', {}, Hub),
            editions: {
                /**
                 * Creates an Edition inside the Event
                 * @param resource The new Edition to create
                 */
                create: (resource) => this.createLinkedResource('create-edition', {}, resource, Edition),
                /**
                 * Retrieves a list of Editions associated with this Event.
                 * @param options Pagination options
                 */
                list: (options) => this.fetchLinkedResource('editions', options, EditionsPage),
            },
            /**
             * Updates this Event with the changes in the mutation parameter.
             * @param mutation Mutated event
             */
            update: (mutation) => this.updateResource(mutation, Event),
        };
    }
}
/**
 * @hidden
 */
export class EventsPage extends Page {
    constructor(data) {
        super('events', Event, data);
    }
}
