// dc-management-sdk-js/build/module/lib/model/Extension.js
import { HalResource } from '../hal/models/HalResource';
import { Hub } from './Hub';
import { Page } from './Page';
/**
 * Class representing the [Extension](https://amplience.com/docs/api/dynamic-content/management/#tag/Extensions) resource.
 */
export class Extension extends HalResource {
    constructor() {
        super(...arguments);
        /**
         * Resources and actions related to an Extension
         */
        this.related = {
            /**
             * Updates this extension with the changes in the mutation parameter.
             */
            update: (mutation) => this.updateResource(mutation, Extension),
            /**
             * Retrieves the Hub this extension is stored in
             */
            hub: () => this.fetchLinkedResource('hub', {}, Hub),
            /**
             * Deletes this extension.
             */
            delete: () => this.deleteResource(),
        };
    }
}
/**
 * @hidden
 */
export class ExtensionsPage extends Page {
    constructor(data) {
        super('extensions', Extension, data);
    }
}
