// dc-management-sdk-js/build/module/lib/model/Folder.js
import { HalResource } from '../hal/models/HalResource';
import { CURIEs } from '../hal/services/CURIEs';
import { ContentItemsPage } from './ContentItem';
import { ContentRepository } from './ContentRepository';
import { Page } from './Page';
export class Folder extends HalResource {
    constructor() {
        super(...arguments);
        /**
         * Resources and actions related to a Folder
         */
        this.related = {
            /**
             * Retrieves the ContentRepository this folder is stored in
             */
            contentRepository: () => this.fetchLinkedResource('content-repository', {}, ContentRepository),
            folders: {
                /**
                 * Retrieves the parent folder of this Folder
                 */
                parent: () => this.fetchLinkedResource('parent-folder', {}, Folder),
                /**
                 * Retrieves the list of sub-folders contained within this Folder
                 * @param options Pagination options
                 */
                list: (options) => {
                    const path = CURIEs.expand(`folders/${this.id}/folders{?page,size,sort}`, options);
                    return this.client.fetchResource(path, FoldersPage);
                },
                /**
                 * Creates a folder beneath the folder
                 * @param resource The new Folder to create
                 */
                create: (resource) => this.createLinkedResource('create-folder', {}, resource, Folder),
            },
            contentItems: {
                /**
                 * Retrieves a list of Content Items stored within this Folder
                 * @param options Pagination options
                 */
                list: (options) => this.fetchLinkedResource('content-items', options, ContentItemsPage),
            },
        };
    }
}
/**
 * @hidden
 */
export class FoldersPage extends Page {
    constructor(data) {
        super('folders', Folder, data);
    }
}
