// dc-management-sdk-js/build/module/lib/model/Hub.js
import { HalResource } from '../hal/models/HalResource';
import { ContentItemsFacets } from './ContentItem';
import { ContentRepositoriesPage, } from './ContentRepository';
import { ContentType, ContentTypePage } from './ContentType';
import { ContentTypeSchema, ContentTypeSchemaPage } from './ContentTypeSchema';
import { EditionsPage } from './Edition';
import { Event, EventsPage } from './Event';
import { Page } from './Page';
import { SearchIndex, SearchIndexesPage } from './SearchIndex';
import { Settings } from './Settings';
import { Webhook, WebhooksPage } from './Webhook';
import { WorkflowState, WorkflowStatesPage } from './WorkflowState';
import { Extension, ExtensionsPage } from './Extension';
import { SnapshotResultList } from './SnapshotResultList';
/**
 * Class representing the [Hub](https://amplience.com/docs/api/dynamic-content/management/#resources-hubs) resource.
 * Hubs are containers for multiple repositories including media, content, content types.
 */
export class Hub extends HalResource {
    constructor() {
        super(...arguments);
        /**
         * Resources and actions related to a Hub
         */
        this.related = {
            contentTypeSchema: {
                /**
                 * Creates a new Content Type Schema
                 * @param resource
                 */
                create: (resource) => this.createLinkedResource('create-content-type-schema', {}, resource, ContentTypeSchema),
                /**
                 * Retrieves a list of Content Type Schemas associated with this Hub
                 * @param options Pagination options
                 */
                list: (options) => this.fetchLinkedResource('list-content-type-schemas', options, ContentTypeSchemaPage),
            },
            contentRepositories: {
                /**
                 * Retrieves a list of Content Repositories associated with this Hub
                 * @param options Pagination options
                 */
                list: (options) => this.fetchLinkedResource('content-repositories', options, ContentRepositoriesPage),
                /**
                 * Find all the Content Repositories that contain a feature
                 * @param options Options
                 */
                findByFeature: (feature, options) => this.client.fetchLinkedResource({
                    href: `hubs/${this.id}/content-repositories/search/findByFeaturesContaining?feature={feature}{&page,size,sort}`,
                    templated: true,
                }, { feature, ...options }, ContentRepositoriesPage),
            },
            contentTypes: {
                /**
                 * Retrieves a list of Content Types associated with this Hub
                 * @param options Pagination options
                 */
                list: (options) => this.fetchLinkedResource('content-types', options, ContentTypePage),
                /**
                 * Register a Content Type on a hub so it can be enabled on a repository
                 */
                register: (resource) => this.createLinkedResource('register-content-type', {}, resource, ContentType),
                /**
                 * Get a content type by its id
                 */
                get: (id) => this.client.fetchResource(`content-types/${id}`, ContentType),
            },
            contentItems: {
                /**
                 * Facet content items and search by text. See [Text Search Query Syntax](https://amplience.com/docs/api/dynamic-content/management/#section/Text-Search-Query-Syntax) for a summary of the query syntax.
                 */
                facet: (facetQuery, options) => this.performActionThatReturnsResource('facet-content-items', options, facetQuery, ContentItemsFacets),
            },
            editions: {
                /**
                 * Find editions by date associated with this Hub
                 * @param options FindByDate & Pageable & Sortable options
                 */
                findByDate: (options, projection) => this.fetchLinkedResource('search-editions', { projection, ...options }, EditionsPage),
            },
            events: {
                /**
                 * Creates an Event inside this Hub
                 * @param resource
                 */
                create: (resource) => this.createLinkedResource('create-event', {}, resource, Event),
                /**
                 * Retrieves a list of Events associated with this Hub
                 * @param options Pagination options
                 */
                list: (options) => this.fetchLinkedResource('events', options, EventsPage),
            },
            extensions: {
                /**
                 * Creates an Extension inside this Hub
                 * @param resource
                 */
                create: (resource) => this.createLinkedResource('create-extension', {}, resource, Extension),
                /**
                 * Get an Extension inside this Hub by its name
                 */
                getByName: (name) => this.fetchLinkedResource('extension-by-name', { name }, Extension),
                /**
                 * Retrieves a list of Extension associated with this Hub
                 * @param options Pagination options
                 */
                list: (options) => this.fetchLinkedResource('extensions', options, ExtensionsPage),
            },
            searchIndexes: {
                /**
                 * Create a search index for a given hub
                 * @param resource
                 */
                create: (resource) => this.createLinkedResource('create-algolia-search-index', {}, resource, SearchIndex),
                /**
                 * Get a search index by its id
                 */
                get: (id) => this.client.fetchResource(`algolia-search/${this.id}/indexes/${id}`, SearchIndex),
                /**
                 * Retrieves a list of search indexes associated with this Hub
                 */
                list: (parentId, projection, options) => this.fetchLinkedResource('algolia-search-indexes', { parentId, projection, ...options }, SearchIndexesPage),
            },
            settings: {
                /**
                 * Updates this hub settings with the changes in the mutation parameter.
                 */
                update: (mutation) => this.updateLinkedResource('update-settings', {
                    method: 'PATCH',
                }, mutation, Settings),
            },
            snapshots: {
                /**
                 * Creates a Snapshot inside this Hub
                 * @param request
                 */
                create: (request) => this.performActionThatReturnsResource('batch-create-snapshots', {}, request, SnapshotResultList),
            },
            webhooks: {
                /**
                 * Creates a Webhook inside this Hub
                 * @param resource
                 */
                create: (resource) => this.createLinkedResource('create-webhook', {}, resource, Webhook),
                /**
                 * Get a Webhook inside this Hub by its id
                 */
                get: (id) => this.client.fetchResource(`hubs/${this.id}/webhooks/${id}`, Webhook),
                /**
                 * Retrieves a list of Webhooks associated with this Hub
                 * @param options Pagination options
                 */
                list: (options) => this.fetchLinkedResource('webhooks', options, WebhooksPage),
            },
            workflowStates: {
                /**
                 * Creates a Workflow State inside this Hub
                 * @param resource
                 */
                create: (resource) => this.createLinkedResource('create-workflow-state', {}, resource, WorkflowState),
                /**
                 * Get a Webhook State insidee this Hub by its id
                 * @param id
                 */
                get: (id) => this.client.fetchResource(`workflow-states/${id}`, WorkflowState),
                /**
                 * Retrieves a list of Workflow States associated with this Hub
                 * @param options Pagination options
                 */
                list: (options) => this.fetchLinkedResource('workflow-states', options, WorkflowStatesPage),
            },
        };
    }
}
/**
 * @hidden
 */
export class HubsPage extends Page {
    constructor(data) {
        super('hubs', Hub, data);
    }
}
