// dc-management-sdk-js/build/module/lib/model/SearchIndex.js
import { HalResource } from '../hal/models/HalResource';
import { AssignedContentType, AssignedContentTypePage, } from './AssignedContentType';
import { Page } from './Page';
import { SearchIndexKey } from './SearchIndexKey';
import { SearchIndexSettings } from './SearchIndexSettings';
import { SearchIndexStatistics } from './SearchIndexStatistics';
import { SearchIndexTopHitsCollection } from './SearchIndexTopHits';
import { SearchIndexTopSearchesCollection, } from './SearchIndexTopSearches';
import { SearchIndexSearchesWithNoResultsCollection } from './SearchIndexSearchesWithNoResults';
import { SearchIndexTopFiltersNoResultSearchCollection } from './SearchIndexTopFiltersNoResultSearch';
import { SearchIndexUsersCount } from './SearchIndexUsersCount';
import { SearchIndexSearchesCount } from './SearchIndexSearchesCount';
import { SearchIndexNoResultsRate } from './SearchIndexNoResultsRate';
import { Hub } from './Hub';
/**
 * Class representing an Algolia Search Index.
 */
export class SearchIndex extends HalResource {
    constructor() {
        super(...arguments);
        /**
         * Resources and actions related to a Search Index.
         */
        this.related = {
            /**
             * Retrieves the Hub this search index is stored in
             */
            hub: () => this.fetchLinkedResource('hub', {}, Hub),
            clear: () => this.performActionThatReturnsResource('clear', {}, {}, SearchIndex),
            delete: () => this.deleteResource(),
            update: (resource) => this.updateResource(resource, SearchIndex),
            assignedContentTypes: {
                create: (resource) => this.createLinkedResource('create-assigned-content-types', {}, resource, AssignedContentType),
                get: (id) => this.client.fetchResource(
                // tslint:disable-next-line:no-string-literal
                `${this._links['self'].href}/assigned-content-types/${id}`, AssignedContentType),
                list: (options) => this.fetchLinkedResource('assigned-content-types', { options }, AssignedContentTypePage),
            },
            indexObject: {
                delete: (id) => 
                // tslint:disable-next-line:no-string-literal
                this.client.deleteResource(`${this._links['self'].href}/objects/${id}`),
            },
            keys: {
                get: () => this.fetchLinkedResource('hub-search-key', {}, SearchIndexKey),
            },
            replicas: {
                list: (projection, options) => this.fetchLinkedResource('list-replicas', { projection, options }, SearchIndexesPage),
            },
            settings: {
                get: () => this.fetchLinkedResource('settings', {}, SearchIndexSettings),
                update: (resource, forwardToReplicas) => this.updateLinkedResource('update-settings', { forwardToReplicas }, resource, SearchIndexSettings),
            },
            stats: {
                get: (period) => this.fetchLinkedResource('stats', { period }, SearchIndexStatistics),
            },
            'top-searches': {
                get: ({ clickAnalytics = false, orderBy, direction, startDate, endDate, limit, offset, tags, includeReplicas, }) => this.fetchLinkedResource('top-searches', {
                    clickAnalytics,
                    direction,
                    endDate,
                    limit,
                    offset,
                    orderBy,
                    startDate,
                    tags,
                    includeReplicas,
                }, SearchIndexTopSearchesCollection),
            },
            'top-hits': {
                get: ({ search, startDate, endDate, limit, offset, tags, includeReplicas, }) => this.fetchLinkedResource('top-hits', {
                    endDate,
                    limit,
                    offset,
                    search,
                    startDate,
                    tags,
                    includeReplicas,
                }, SearchIndexTopHitsCollection),
            },
            'searches-with-no-results': {
                get: ({ startDate, endDate, limit, offset, tags, includeReplicas, }) => this.fetchLinkedResource('searches-with-no-results', {
                    endDate,
                    limit,
                    offset,
                    startDate,
                    tags,
                    includeReplicas,
                }, SearchIndexSearchesWithNoResultsCollection),
            },
            'top-filters-no-result-search': {
                get: ({ search, startDate, endDate, limit, offset, tags, includeReplicas, }) => this.fetchLinkedResource('top-filters-no-result-search', {
                    endDate,
                    limit,
                    offset,
                    search,
                    startDate,
                    tags,
                    includeReplicas,
                }, SearchIndexTopFiltersNoResultSearchCollection),
            },
            'users-count': {
                get: ({ startDate, endDate, tags, includeReplicas, }) => this.fetchLinkedResource('users-count', {
                    startDate,
                    endDate,
                    tags,
                    includeReplicas,
                }, SearchIndexUsersCount),
            },
            'searches-count': {
                get: ({ startDate, endDate, tags, includeReplicas, }) => this.fetchLinkedResource('searches-count', {
                    startDate,
                    endDate,
                    tags,
                    includeReplicas,
                }, SearchIndexSearchesCount),
            },
            'no-results-rate': {
                get: ({ startDate, endDate, tags, includeReplicas, }) => this.fetchLinkedResource('no-results-rate', {
                    startDate,
                    endDate,
                    tags,
                    includeReplicas,
                }, SearchIndexNoResultsRate),
            },
        };
    }
}
/**
 * @hidden
 */
export class SearchIndexesPage extends Page {
    constructor(data) {
        super('indexes', SearchIndex, data);
    }
}
