// dc-management-sdk-js/build/module/lib/model/Webhook.js
import { HalResource } from '../hal/models/HalResource';
import { Hub } from './Hub';
import { Page } from './Page';
/**
 * Class representing the [Webhook](https://amplience.com/docs/api/dynamic-content/management/#tag/Webhooks) resource.
 * A Webhook is a HTTP callback: a HTTP request that occurs when something happens; a simple event-notification via the HTTP protocol.
 */
export class Webhook extends HalResource {
    constructor() {
        super(...arguments);
        /**
         * Resources and actions related to a Webhook
         */
        this.related = {
            /**
             * Retrieves the Hub this webhook is stored in
             */
            hub: () => this.fetchLinkedResource('hub', {}, Hub),
            /**
             * Delete Webhook
             */
            delete: () => this.deleteResource(),
            /**
             * Updates this Webhook with the changes in the mutation parameter.
             * @param mutation
             */
            update: (mutation) => this.updateResource(mutation, Webhook),
        };
    }
    parse(data) {
        super.parse(data);
        if (this.headers) {
            this.headers.forEach((header) => {
                // secret can come back as null, lets convert this to false
                if (header.secret == null) {
                    header.secret = false;
                }
            });
        }
    }
}
/**
 * @hidden
 */
export class WebhooksPage extends Page {
    constructor(data) {
        super('webhooks', Webhook, data);
    }
}
