// dc-management-sdk-js/build/module/lib/hal/utils/HalMock.js
import { CURIEs } from '../services/CURIEs';
/**
 * @hidden
 */
export class HalMockResource {
    constructor(resource, mocks) {
        this.resource = resource;
        this.mocks = mocks;
    }
    nestedResource(linkName, args, resource) {
        const link = this.resource._links[linkName];
        const href = CURIEs.expand(link.href, args);
        this.mocks.resource(resource, href);
        return this;
    }
    nestedCollection(linkName, args, type, values) {
        const link = this.resource._links[linkName];
        const href = CURIEs.expand(link.href, args);
        this.mocks.collection(href, type, values);
        return this;
    }
    nestedCreateResource(linkName, args, resource, errorCode = 200) {
        const link = this.resource._links[linkName];
        const href = CURIEs.expand(link.href, args);
        this.mocks.createResource(href, resource, errorCode);
        return this;
    }
    nestedPutResource(linkName, args, resource) {
        const link = this.resource._links[linkName];
        const href = CURIEs.expand(link.href, args);
        this.mocks.putResource(href, resource);
        return this;
    }
    nestedUpdateResource(linkName, args, resource) {
        const link = this.resource._links[linkName];
        const href = CURIEs.expand(link.href, args);
        this.mocks.updateResource(href, resource);
        return this;
    }
    nestedDelete(linkName, args) {
        const link = this.resource._links[linkName];
        const href = CURIEs.expand(link.href, args);
        this.mocks.deleteResource(href);
        return this;
    }
}
/**
 * @hidden
 */
export class HalMocks {
    constructor(mockInstance) {
        this.mockInstance = mockInstance;
    }
    resource(resource, url) {
        if (!url) {
            url = resource._links.self.href;
        }
        this.mockInstance.onGet(url).reply(200, resource);
        if (resource._links.delete) {
            this.mockInstance.onDelete(resource._links.delete.href).reply(204);
        }
        return new HalMockResource(resource, this);
    }
    collection(url, type, values) {
        this.mockInstance.onGet(url).reply(200, {
            _embedded: {
                [type]: values,
            },
        });
    }
    createResource(url, resource, errorCode = 200) {
        this.mockInstance.onPost(url).reply(errorCode, resource);
        return new HalMockResource(resource, this);
    }
    putResource(url, resource) {
        this.mockInstance.onPut(url).reply(200, resource);
        return new HalMockResource(resource, this);
    }
    updateResource(url, resource) {
        this.mockInstance.onPatch(url).reply(200, resource);
        return new HalMockResource(resource, this);
    }
    deleteResource(url) {
        this.mockInstance.onDelete(url).reply(200);
    }
    postResource(url) {
        this.mockInstance.onPost(url).reply(204);
    }
}
